{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GTEx data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tr = transcript ratio;  ec = exon correlation;\n",
    "formatted_tr = \"/home/lzg/1906/QTL_data/unfiltered/sQTL_transcript.txt\"\n",
    "raw_dir = \"/home/galaxy/project/m6AQTL/data/sqtl/total_data/GTEx\"\n",
    "trans_ratio = \"%s/sQTLSeekeR_Transcript-Ratio/sQTLSeekeR_Transcript-Ratio.csv\" % raw_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# process trans_ratio\n",
    "def generate_tr_dict():\n",
    "    df = pd.read_csv(trans_ratio)\n",
    "    df[\"query_id\"] = df[\"Gene Symbol\"].astype(str) + \";\" + df[\"SNP\"].astype(str)\n",
    "    query_list = [x.strip() for x in df[\"query_id\"].tolist()] # remove the end blank of SNP\n",
    "    df[\"value\"] = \"Transcript-Ratio\" + \";\" + df[\"Transcript 1\"].str.split(\".\").str[0] + \";\" + df[\"Transcript 2\"].str.split(\".\").str[0]\n",
    "    tr_dict = dict(zip(query_list, df[\"value\"]))\n",
    "    print(tr_dict[\"C1orf201;rs1016209\"]) # supposed to print out Transcript-Ratio;ENST00000337248.4;ENST00000003583.8\n",
    "    return tr_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_ensembl_coordinate():\n",
    "    gff = \"/data/database/GRCh38/GENCODE/gencode.v27.annotation.gff3\"\n",
    "    df = pd.read_table(gff, sep=\"\\t\", comment=\"#\", header=None)\n",
    "    df_tc = df[df.iloc[:, 2] == \"transcript\"]\n",
    "    df_tc[\"Ensembl_ID\"] = df_tc.iloc[:, 8].str.split(\"ID=\").str[1].str.split(\".\").str[0]\n",
    "    df_tc[\"coor\"] = df_tc.iloc[:, 0] + \":\" + df_tc.iloc[:, 3].astype(str) + \"-\" + df_tc.iloc[:, 4].astype(str)\n",
    "    ens_coor_dict = dict(zip(df_tc[\"Ensembl_ID\"], df_tc[\"coor\"]))\n",
    "    print(\"\\t\" + ens_coor_dict[\"ENST00000456328\"]) # supposed to be chr1:11869-14409\n",
    "    return ens_coor_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_id_for_TranscriptRatio():\n",
    "    dict_tr = generate_tr_dict()\n",
    "    ens_coor_dict = generate_ensembl_coordinate()\n",
    "    #\n",
    "    df_tr = pd.read_table(formatted_tr, sep=\"\\t\", skiprows=1)\n",
    "    df_tr[\"query_id\"] = df_tr[\"Gene_Name\"].astype(str) + \";\" + df_tr[\"RS_ID\"].astype(str)\n",
    "    query_list =[x.strip() for x in df_tr[\"query_id\"].tolist()] # remove the end blank space;\n",
    "    id_list = [dict_tr[x] if x in dict_tr else str(np.nan) for x in query_list]\n",
    "    coor_list = []\n",
    "    for sp_id in id_list:\n",
    "        pre = sp_id.split(\";\")[0]\n",
    "        post = \";\".join([ens_coor_dict[x] if x in ens_coor_dict else str(np.nan) for x in sp_id.split(\";\")[1:]])\n",
    "        coor_list.append(pre + \";\" + post)\n",
    "    df_tr[\"Splicing_ID\"] = coor_list\n",
    "    # can't decide sQTL type\n",
    "    df_tr[\"sQTL_type\"] = \"cis\"\n",
    "    df_tr[\"Cutoff\"] = \"5kb upstream or downstream of each gene's transcript start or end sites\"\n",
    "    del df_tr['query_id']\n",
    "    df_tr.to_csv(\"/home/galaxy/project/m6AQTL/data/sqtl/total_data/sQTL_TranscriptRatio.txt\", sep=\"\\t\", index=False)\n",
    "    print(df_tr.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcript-Ratio;ENST00000337248;ENST00000003583\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/galaxy/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"\n",
      "/home/galaxy/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tchr1:11869-14409\n",
      "        Ensembl_ID Gene_Name  Position Splicing_ID  \\\n",
      "0  ENSG00000001460     STPG1       NaN        nan;   \n",
      "1  ENSG00000001460     STPG1       NaN        nan;   \n",
      "2  ENSG00000001460     STPG1       NaN        nan;   \n",
      "3  ENSG00000001460     STPG1       NaN        nan;   \n",
      "4  ENSG00000001460     STPG1       NaN        nan;   \n",
      "\n",
      "                                        Event_type        RS_ID  REF  ALT  \\\n",
      "0                                    exon skipping   rs1016209   NaN  NaN   \n",
      "1                                    exon skipping  rs11589294   NaN  NaN   \n",
      "2                                    exon skipping  rs11587667   NaN  NaN   \n",
      "3                                    exon skipping   rs7529526   NaN  NaN   \n",
      "4  complex event 5';complex event;complex event 3'  rs12140084   NaN  NaN   \n",
      "\n",
      "  sQTL_type                                             Cutoff    Pvalue  \\\n",
      "0       cis  5kb upstream or downstream of each gene's tran...  0.000024   \n",
      "1       cis  5kb upstream or downstream of each gene's tran...  0.000028   \n",
      "2       cis  5kb upstream or downstream of each gene's tran...  0.000028   \n",
      "3       cis  5kb upstream or downstream of each gene's tran...  0.000028   \n",
      "4       cis  5kb upstream or downstream of each gene's tran...  0.000032   \n",
      "\n",
      "   Corrected_pvalue         Measure  Value        Cell_or_tissue  Condition  \\\n",
      "0           0.00985  Max Difference  0.131  Adipose_Subcutaneous        NaN   \n",
      "1           0.01120  Max Difference  0.131  Adipose_Subcutaneous        NaN   \n",
      "2           0.01120  Max Difference  0.131  Adipose_Subcutaneous        NaN   \n",
      "3           0.01120  Max Difference  0.131  Adipose_Subcutaneous        NaN   \n",
      "4           0.01220  Max Difference  0.187  Adipose_Subcutaneous        NaN   \n",
      "\n",
      "   Experimental_tag  PubMed_ID           query_id  \n",
      "0                 1   25954001   STPG1;rs1016209   \n",
      "1                 1   25954001  STPG1;rs11589294   \n",
      "2                 1   25954001  STPG1;rs11587667   \n",
      "3                 1   25954001   STPG1;rs7529526   \n",
      "4                 1   25954001  STPG1;rs12140084   \n"
     ]
    }
   ],
   "source": [
    "add_id_for_TranscriptRatio()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl_ID</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>Position</th>\n",
       "      <th>Splicing_ID</th>\n",
       "      <th>Event_type</th>\n",
       "      <th>RS_ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>sQTL_type</th>\n",
       "      <th>Cutoff</th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>Corrected_pvalue</th>\n",
       "      <th>Measure</th>\n",
       "      <th>Value</th>\n",
       "      <th>Cell_or_tissue</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Experimental_tag</th>\n",
       "      <th>PubMed_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000116906</td>\n",
       "      <td>GNPAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:231403426-231403649;chr1:231409668-231409808</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs10779813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cis</td>\n",
       "      <td>1M</td>\n",
       "      <td>9.810000e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spearman Rho</td>\n",
       "      <td>0.449</td>\n",
       "      <td>Artery Tibial</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>25954001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr20:49574900-49575058;chr20:49562274-49562299</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs73130786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cis</td>\n",
       "      <td>1M</td>\n",
       "      <td>2.090000e-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spearman Rho</td>\n",
       "      <td>-0.446</td>\n",
       "      <td>Thyroid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>25954001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:169822215-169822913;chr1:169823411-169824105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs11585877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cis</td>\n",
       "      <td>1M</td>\n",
       "      <td>9.010000e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spearman Rho</td>\n",
       "      <td>-0.450</td>\n",
       "      <td>Artery Tibial</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>25954001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:169831754-169831938;chr1:169822215-169822913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs10753798</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cis</td>\n",
       "      <td>1M</td>\n",
       "      <td>1.170000e-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spearman Rho</td>\n",
       "      <td>-0.473</td>\n",
       "      <td>Skin Sun Exposed (Lower leg)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>25954001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:196658550-196658744;chr1:196654194-196654367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs10921983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cis</td>\n",
       "      <td>1M</td>\n",
       "      <td>8.480000e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spearman Rho</td>\n",
       "      <td>0.482</td>\n",
       "      <td>Adipose Subcutaneous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>25954001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Ensembl_ID Gene_Name  Position  \\\n",
       "0  ENSG00000116906     GNPAT       NaN   \n",
       "1  ENSG00000000419      DPM1       NaN   \n",
       "2  ENSG00000000457     SCYL3       NaN   \n",
       "3  ENSG00000000457     SCYL3       NaN   \n",
       "4  ENSG00000000971       CFH       NaN   \n",
       "\n",
       "                                         Splicing_ID  Event_type       RS_ID  \\\n",
       "0  chr1:231403426-231403649;chr1:231409668-231409808         NaN  rs10779813   \n",
       "1    chr20:49574900-49575058;chr20:49562274-49562299         NaN  rs73130786   \n",
       "2  chr1:169822215-169822913;chr1:169823411-169824105         NaN  rs11585877   \n",
       "3  chr1:169831754-169831938;chr1:169822215-169822913         NaN  rs10753798   \n",
       "4  chr1:196658550-196658744;chr1:196654194-196654367         NaN  rs10921983   \n",
       "\n",
       "   REF  ALT sQTL_type Cutoff        Pvalue  Corrected_pvalue       Measure  \\\n",
       "0  NaN  NaN       cis     1M  9.810000e-07               NaN  Spearman Rho   \n",
       "1  NaN  NaN       cis     1M  2.090000e-06               NaN  Spearman Rho   \n",
       "2  NaN  NaN       cis     1M  9.010000e-07               NaN  Spearman Rho   \n",
       "3  NaN  NaN       cis     1M  1.170000e-06               NaN  Spearman Rho   \n",
       "4  NaN  NaN       cis     1M  8.480000e-07               NaN  Spearman Rho   \n",
       "\n",
       "   Value                Cell_or_tissue  Condition  Experimental_tag  PubMed_ID  \n",
       "0  0.449                 Artery Tibial        NaN                 2   25954001  \n",
       "1 -0.446                       Thyroid        NaN                 8   25954001  \n",
       "2 -0.450                 Artery Tibial        NaN                 2   25954001  \n",
       "3 -0.473  Skin Sun Exposed (Lower leg)        NaN                 7   25954001  \n",
       "4  0.482          Adipose Subcutaneous        NaN                10   25954001  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process ExonCorrelation\n",
    "ec = \"/home/lzg/1906/QTL_data/unfiltered/sQTL_exon.txt\"\n",
    "df = pd.read_table(ec, sep=\"\\t\", skiprows=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_beds = df[\"Splicing_ID\"].tolist()\n",
    "raw_list = []\n",
    "for i in range(len(raw_beds)):\n",
    "    a, b = raw_beds[i].split(\";\")[0], raw_beds[i].split(\";\")[1]\n",
    "    a_1 = \"\\t\".join([a.split(\":\")[0], str(a.split(\":\")[1].split(\"-\")[0]), str(a.split(\"-\")[1]), str(i)+\".\"+str(1)])\n",
    "    b_2 = \"\\t\".join([b.split(\":\")[0], str(b.split(\":\")[1].split(\"-\")[0]), str(b.split(\"-\")[1]), str(i)+\".\"+str(2)])\n",
    "    raw_list.append(a_1)\n",
    "    raw_list.append(b_2)\n",
    "grch37_bed = \"/home/galaxy/project/m6AQTL/data/sqtl/total_data/liftover/grch37.bed\"\n",
    "with open(grch37_bed, 'w') as fw:\n",
    "    fw.writelines([\"%s\\n\" % x for x in raw_list])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ### liftover\n",
    "chain_file, failure_file = \"/data/database/chain_file/hg19ToHg38.over.chain\", \"/home/galaxy/failure.txt\"\n",
    "grch37_bed = \"/home/galaxy/project/m6AQTL/data/sqtl/total_data/liftover/grch37.bed\"\n",
    "grch38_bed = \"/home/galaxy/project/m6AQTL/data/sqtl/total_data/liftover/grch38.bed\"\n",
    "command = \"liftOver %s %s %s %s\" % (grch37_bed, chain_file, grch38_bed, failure_file)\n",
    "os.system(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41185\n",
      "  chr_name      start        end  line_num\n",
      "0     chr1  231267680  231267903       0.1\n",
      "1     chr1  231273922  231274062       0.2\n",
      "2    chr20   50958363   50958521       1.1\n",
      "3    chr20   50945737   50945762       1.2\n",
      "4     chr1  169853074  169853772       2.1\n",
      "      chr_name       start         end  line_num                      id\n",
      "82286    chr16   4407562.0   4407684.0   41182.2   chr16:4407562-4407684\n",
      "82287    chr19   7470126.0   7470245.0   41183.1   chr19:7470126-7470245\n",
      "82288    chr19   7476875.0   7477436.0   41183.2   chr19:7476875-7477436\n",
      "82289     chr6  31669118.0  31669172.0   41184.1  chr6:31669118-31669172\n",
      "82290     chr6  31665861.0  31666283.0   41184.2  chr6:31665861-31666283\n"
     ]
    }
   ],
   "source": [
    "line_num = len(df)\n",
    "print(line_num)\n",
    "df_38 = pd.read_table(grch38_bed, sep=\"\\t\", header=None)\n",
    "df_38.columns = [\"chr_name\", \"start\", \"end\", \"line_num\"]\n",
    "print(df_38.head())\n",
    "full_list = []\n",
    "for i in range(line_num):\n",
    "    full_list.append([i+0.1])\n",
    "    full_list.append([i+0.2])\n",
    "df_full = pd.DataFrame(full_list)\n",
    "df_full.columns = [\"line_num\"]\n",
    "df_outer = df_38.merge(df_full, on=\"line_num\", how=\"right\")\n",
    "df_outer = df_outer.sort_values([\"line_num\"])\n",
    "df_outer[\"id\"] = df_outer[\"chr_name\"] + \":\" + df_outer[\"start\"].astype(str).str.split(\".\").str[0] + \"-\" + df_outer[\"end\"].astype(str).str.split(\".\").str[0]\n",
    "print(df_outer.tail())\n",
    "id_list = df_outer[\"id\"].tolist()\n",
    "result_list = []\n",
    "for i in range(0, len(id_list)-1, 2):\n",
    "    result_list.append(\";\".join([\"ExonCorrelation\", str(id_list[i]), str(id_list[i+1])]))\n",
    "df_1 = df\n",
    "assert len(df_1[\"Splicing_ID\"]) == len(result_list)\n",
    "df_1[\"Splicing_ID\"] = result_list\n",
    "df_1.to_csv(\"/home/galaxy/project/m6AQTL/data/sqtl/total_data/sQTL_ExonCorrelation.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### process paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20417\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl_ID</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>Position</th>\n",
       "      <th>Splicing_ID</th>\n",
       "      <th>Event_type</th>\n",
       "      <th>RS_ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>sQTL_type</th>\n",
       "      <th>Cutoff</th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>Corrected_pvalue</th>\n",
       "      <th>Measure</th>\n",
       "      <th>Value</th>\n",
       "      <th>Cell_or_tissue</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Experimental_tag</th>\n",
       "      <th>PubMed_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000143337</td>\n",
       "      <td>TOR1AIP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:179884769-179889310:chr1:179884769-179889313</td>\n",
       "      <td>A3</td>\n",
       "      <td>rs2245425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.611478e-69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>peripheral blood mononuclear cells(PBMCs)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30975994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000181381</td>\n",
       "      <td>DDX60L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr4:168396127-168400826:chr4:168396124-168400826</td>\n",
       "      <td>A3</td>\n",
       "      <td>rs3857035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.642871e-69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>peripheral blood mononuclear cells(PBMCs)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30975994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000181381</td>\n",
       "      <td>DDX60L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr4:168375524-168377728:chr4:168377843-168378354</td>\n",
       "      <td>SE</td>\n",
       "      <td>rs2046770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.451629e-31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>peripheral blood mononuclear cells(PBMCs)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30975994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000181381</td>\n",
       "      <td>DDX60L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr4:168472809-168475996:chr4:168476120-168480377</td>\n",
       "      <td>SE</td>\n",
       "      <td>rs4692942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.732243e-18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>peripheral blood mononuclear cells(PBMCs)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30975994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000181381</td>\n",
       "      <td>DDX60L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr4:168461930-168461931:chr4:168462040-168471...</td>\n",
       "      <td>AL</td>\n",
       "      <td>rs56999040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.147239e-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>peripheral blood mononuclear cells(PBMCs)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30975994.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Ensembl_ID Gene_Name  Position  \\\n",
       "0  ENSG00000143337  TOR1AIP1       NaN   \n",
       "1  ENSG00000181381    DDX60L       NaN   \n",
       "2  ENSG00000181381    DDX60L       NaN   \n",
       "3  ENSG00000181381    DDX60L       NaN   \n",
       "4  ENSG00000181381    DDX60L       NaN   \n",
       "\n",
       "                                         Splicing_ID Event_type       RS_ID  \\\n",
       "0  chr1:179884769-179889310:chr1:179884769-179889313         A3   rs2245425   \n",
       "1  chr4:168396127-168400826:chr4:168396124-168400826         A3   rs3857035   \n",
       "2  chr4:168375524-168377728:chr4:168377843-168378354         SE   rs2046770   \n",
       "3  chr4:168472809-168475996:chr4:168476120-168480377         SE   rs4692942   \n",
       "4  chr4:168461930-168461931:chr4:168462040-168471...         AL  rs56999040   \n",
       "\n",
       "   REF  ALT sQTL_type Cutoff        Pvalue  Corrected_pvalue Measure  Value  \\\n",
       "0  NaN  NaN       NaN    NaN  1.611478e-69               NaN     NaN    NaN   \n",
       "1  NaN  NaN       NaN    NaN  2.642871e-69               NaN     NaN    NaN   \n",
       "2  NaN  NaN       NaN    NaN  2.451629e-31               NaN     NaN    NaN   \n",
       "3  NaN  NaN       NaN    NaN  1.732243e-18               NaN     NaN    NaN   \n",
       "4  NaN  NaN       NaN    NaN  3.147239e-08               NaN     NaN    NaN   \n",
       "\n",
       "                              Cell_or_tissue  Condition  Experimental_tag  \\\n",
       "0  peripheral blood mononuclear cells(PBMCs)        NaN               1.0   \n",
       "1  peripheral blood mononuclear cells(PBMCs)        NaN               1.0   \n",
       "2  peripheral blood mononuclear cells(PBMCs)        NaN               1.0   \n",
       "3  peripheral blood mononuclear cells(PBMCs)        NaN               1.0   \n",
       "4  peripheral blood mononuclear cells(PBMCs)        NaN               1.0   \n",
       "\n",
       "    PubMed_ID  \n",
       "0  30975994.0  \n",
       "1  30975994.0  \n",
       "2  30975994.0  \n",
       "3  30975994.0  \n",
       "4  30975994.0  "
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_file = \"/home/lzg/1906/QTL_data/unfiltered/sQTL_paper.txt\"\n",
    "df = pd.read_table(raw_file, sep=\"\\t\", skiprows=1)\n",
    "print(len(df))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19700\n",
      "  chr_name        start          end  line_num                        id\n",
      "0     chr1  179915634.0  179920175.0       1.1  chr1:179915634-179920175\n",
      "1     chr1  179915634.0  179920178.0       1.2  chr1:179915634-179920178\n",
      "2     chr4  167474976.0  167479675.0       2.1  chr4:167474976-167479675\n",
      "3     chr4  167474973.0  167479675.0       2.2  chr4:167474973-167479675\n",
      "4     chr4  167454373.0  167456577.0       3.1  chr4:167454373-167456577\n"
     ]
    }
   ],
   "source": [
    "#########################################################################################################\n",
    "chain_file, failure_file = \"/data/database/chain_file/hg19ToHg38.over.chain\", \"/home/galaxy/failure.txt\"\n",
    "raw_bed = \"/home/galaxy/project/m6AQTL/data/sqtl/total_data/liftover/grch37.bed\"\n",
    "final_bed = \"/home/galaxy/project/m6AQTL/data/sqtl/total_data/liftover/grch38.bed\"\n",
    "id_list = df[\"Splicing_ID\"].tolist()\n",
    "result_file = \"/home/galaxy/project/m6AQTL/data/sqtl/total_data/sQTL_paper.txt\"\n",
    "# replace hg19 coordinates with hg38;\n",
    "#########################################################################################################\n",
    "\n",
    "\n",
    "def map_to_original():\n",
    "    df_38 = pd.read_table(final_bed, sep=\"\\t\", header=None)\n",
    "    df_38.columns = [\"chr_name\", \"start\", \"end\", \"line_num\"]\n",
    "    df_full = pd.DataFrame(\n",
    "        pd.read_table(raw_bed, sep=\"\\t\", header=None).iloc[:, -1])\n",
    "    df_full.columns = [\"line_num\"]\n",
    "    df_outer = df_38.merge(df_full, on=\"line_num\", how=\"right\")\n",
    "    df_outer[\"id\"] = df_outer[\"chr_name\"] + \":\" + df_outer[\"start\"].astype(\n",
    "        str).str.split(\".\").str[0] + \"-\" + df_outer[\"end\"].astype(\n",
    "            str).str.split(\".\").str[0]\n",
    "    print(df_outer.head())\n",
    "    id_list = df_outer[\"id\"].tolist()\n",
    "    line_nums = df_outer[\"line_num\"].tolist()\n",
    "    result_list = []\n",
    "    for i in range(1, len(df)+1):\n",
    "        result_list.append(\";\".join([\n",
    "            str(id_list[j]) for j in range(len(id_list))\n",
    "            if str(line_nums[j]).startswith(str(i) + \".\")\n",
    "        ]))\n",
    "    df_1 = df\n",
    "    assert len(df_1[\"Splicing_ID\"]) == len(result_list)\n",
    "    df_1[\"Splicing_ID\"] = result_list\n",
    "    #\n",
    "    df_1[\"Experimental_tag\"] = df_1[\"Experimental_tag\"].astype(int)\n",
    "    df_1[\"PubMed_ID\"] = df_1[\"PubMed_ID\"].astype(int)\n",
    "    df_1.to_csv(result_file, sep=\"\\t\", index=False)\n",
    "\n",
    "\n",
    "def liftover():\n",
    "    command = \"liftOver %s %s %s %s\" % (raw_bed, chain_file, final_bed, failure_file)\n",
    "    os.system(command)\n",
    "\n",
    "\n",
    "def add_peakNumber(query_list, result_file):\n",
    "    result_list = []\n",
    "    for i in range(len(query_list)):\n",
    "        query = query_list[i]\n",
    "        if str(query).startswith(\"chr\"):\n",
    "            raw_list = add_peak_number(query, i + 1)\n",
    "            result_list += raw_list\n",
    "    print(len(result_list))\n",
    "    with open(result_file, 'w') as fw:\n",
    "        fw.writelines([\"%s\\n\" % x for x in result_list])\n",
    "\n",
    "\n",
    "def add_peak_number(line, line_num):\n",
    "    #     a = \"chr10:3103945-3105114\"\n",
    "    #     b = \"chr9:137428616-137428882\"\n",
    "    #     c = \"chr12:70344228-70346180:chr12:70346625-70346626:chr12:70344228-70353829:chr12:70353912-70353913\"\n",
    "    peak_list = [\"chr%s\" % x.strip(\":\") for x in line.split(\"chr\")[1:]]\n",
    "    result_list, decimal = [], 0.1\n",
    "    for i in range(len(peak_list)):\n",
    "        peak_num = str(line_num + decimal * (i + 1))\n",
    "        # separator - or +\n",
    "        bed = \"\\t\".join([\n",
    "            peak_list[i].split(\":\")[0],\n",
    "            str(peak_list[i].split(\":\")[1].split(\"-\")[0]),\n",
    "            str(peak_list[i].replace(\"+\", \"-\").split(\"-\")[1].split(\":\")[0]),\n",
    "            peak_num\n",
    "        ])\n",
    "        result_list.append(bed)\n",
    "    return result_list\n",
    "\n",
    "\n",
    "add_peakNumber(id_list, raw_bed)\n",
    "liftover()\n",
    "map_to_original()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
