{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bedtools intersect -a AMR_hg19.bed -b /home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19_exon.bed -wa | sort -u > exon_AMR.bed\n",
    "# bedtools shuffle -noOverlapping -i AMR.bed -g /home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/chrNameLength.txt > Control.bed\n",
    "# bedtools shuffle -noOverlapping -i exon_AMR.bed -incl /home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19_exon.bed -g /home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/chrNameLength.txt > exon_Control.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15274 15274\n",
      "CRR055537.bed\t0.6029411764705886\t2.5714285714285716\t6\t1\t7\t3\n",
      "CRR042308.bed\t1.0\t0.0\t8\t1\t8\t0\n",
      "CRR042318.bed\t0.7856784817721821\t1.2190476190476192\t48\t7\t45\t8\n",
      "CRR055539.bed\t0.41666666666666724\tinf\t7\t0\t4\t1\n",
      "CRR042296.bed\t1.0\t0.9705882352941176\t11\t2\t17\t3\n",
      "CRR073018.bed\t1.0\t0.8\t8\t1\t10\t1\n",
      "CRR055533.bed\t1.0\tnan\t3\t0\t1\t0\n",
      "CRR042280.bed\t0.4388266018561261\t0.4222222222222222\t19\t6\t15\t2\n",
      "CRR042310.bed\t1.0\t0.5\t9\t2\t9\t1\n",
      "CRR055541.bed\t1.0\tnan\t2\t0\t3\t0\n",
      "CRR055543.bed\t0.46428571428571397\t5.0\t5\t1\t1\t1\n",
      "brain_IP_1.bed\t0.7721229946834149\t1.4\t63\t5\t72\t8\n",
      "mus_IP_4.bed\t1.0\t0.5714285714285714\t16\t2\t14\t1\n",
      "CRR055536.bed\t1.0\t0.0\t7\t1\t8\t0\n",
      "sto_IP_5.bed\t0.23913043478260831\tinf\t21\t0\t22\t3\n",
      "brain_IP_3.bed\t0.21935935267000445\t2.0454545454545454\t60\t8\t44\t12\n",
      "CRR055527.bed\t1.0\t1.0416666666666667\t25\t4\t30\t5\n",
      "CRR042288.bed\t1.0\tnan\t5\t0\t7\t0\n",
      "CRR042292.bed\t0.632660778435026\t0.34375\t22\t4\t16\t1\n",
      "CRR042284.bed\t0.7054083100293984\t1.2222222222222223\t55\t18\t50\t20\n",
      "CRR042278.bed\t0.36299921675645863\t1.9487179487179487\t38\t4\t39\t8\n",
      "CRR055563.bed\t1.0\tnan\t4\t0\t2\t0\n",
      "CRR055547.bed\t1.0\t0.0\t5\t1\t2\t0\n",
      "sto_IP_4.bed\t0.4549205297583099\t0.475\t38\t5\t32\t2\n",
      "CRR042298.bed\t1.0\t2.2222222222222223\t10\t1\t9\t2\n",
      "CRR042320.bed\t1.0\tinf\t3\t0\t3\t1\n",
      "liver_IP_1.bed\t0.2787819799777531\t5.4\t18\t1\t10\t3\n",
      "CRR042290.bed\t1.0\t2.0\t15\t1\t15\t2\n",
      "CRR042302.bed\t1.0\t2.5\t10\t1\t4\t1\n",
      "CRR042314.bed\t0.47058823529411775\t0.0\t7\t1\t9\t0\n",
      "liver_IP_2.bed\t0.42586773173302994\t2.5961538461538463\t27\t2\t26\t5\n",
      "CRR055561.bed\t1.0\t1.0909090909090908\t12\t1\t11\t1\n",
      "CRR042294.bed\t1.0\t0.4583333333333333\t11\t2\t12\t1\n",
      "CRR055553.bed\t1.0\t0.0\t14\t1\t12\t0\n",
      "CRR042312.bed\t1.0\t1.8\t9\t1\t10\t2\n",
      "CRR055549.bed\t1.0\tinf\t8\t0\t8\t1\n",
      "CRR042282.bed\t0.4702731496070987\t2.074074074074074\t16\t2\t27\t7\n",
      "CRR055531.bed\t0.17931034482758618\tinf\t17\t0\t11\t2\n",
      "CRR042316.bed\t0.28303432810074175\t5.181818181818182\t19\t1\t11\t3\n",
      "lung_IP_5.bed\t1.0\t0.696969696969697\t23\t3\t22\t2\n",
      "CRR055559.bed\t1.0\tnan\t4\t0\t1\t0\n",
      "pla_IP_4.bed\t0.7894818746133514\t0.8409090909090909\t37\t8\t44\t8\n",
      "brain_IP_2.bed\t0.04814581921001809\t5.454545454545454\t24\t2\t22\t10\n",
      "heart_IP_1.bed\t0.2813612536091731\t0.0\t26\t4\t17\t0\n",
      "CRR073020.bed\t1.0\tnan\t3\t0\t5\t0\n",
      "CRR042300.bed\t1.0\t0.5\t12\t2\t12\t1\n",
      "heart_IP_3.bed\t1.0\t0.9666666666666667\t29\t6\t30\t6\n",
      "pla_IP_2.bed\t0.7758656798077506\t1.3333333333333333\t46\t6\t46\t8\n",
      "CRR055555.bed\t0.42857142857142844\t0.0\t2\t1\t4\t0\n",
      "pla_IP_6.bed\t0.04596440116059858\t0.225\t24\t8\t40\t3\n",
      "CRR073016.bed\t1.0\t0.9333333333333333\t14\t2\t15\t2\n",
      "kid_IP_2.bed\t0.22194379773451514\t1.948051948051948\t75\t6\t77\t12\n",
      "heart_IP_2.bed\t0.3319631146648061\t2.2421875\t41\t4\t32\t7\n",
      "liver_IP_3.bed\t0.4427214837095858\t0.51\t17\t4\t25\t3\n",
      "mus_IP_5.bed\t0.6784839332702467\t2.04\t51\t2\t50\t4\n",
      "kid_IP_4.bed\t0.7865510190627121\t1.1587837837837838\t49\t8\t37\t7\n",
      "CRR055525.bed\t0.7564601623649696\t0.6944444444444444\t35\t7\t36\t5\n",
      "CRR042306.bed\t1.0\tinf\t5\t0\t6\t1\n",
      "CRR042304.bed\t0.8070246371465397\t0.8517857142857143\t53\t10\t56\t9\n",
      "CRR055529.bed\t0.19271516496821675\t0.17037037037037037\t23\t5\t27\t1\n",
      "CRR055557.bed\t1.0\t0.0\t6\t1\t2\t0\n",
      "kid_IP_3.bed\t0.7475667951271667\t1.3295454545454546\t39\t4\t44\t6\n",
      "CRR055551.bed\t1.0\tnan\t3\t0\t5\t0\n",
      "lung_IP_4.bed\t1.0\t1.0731707317073171\t44\t9\t41\t9\n",
      "CRR055545.bed\t0.5714285714285714\t2.75\t11\t1\t8\t2\n",
      "CRR042286.bed\t0.19687602582339495\t0.2\t17\t5\t17\t1\n"
     ]
    }
   ],
   "source": [
    "### allele-specific m6A region\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/results/\")\n",
    "# amr_dir = \"ASM/corr_m6A/corr_results/plot_profile/m6A_region/\"\n",
    "# amr = os.path.join(amr_dir, \"ASM.bed2\")\n",
    "amr_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/AMR_db/\"\n",
    "amr = os.path.join(amr_dir, \"AMR_hg19.bed\")\n",
    "amr_num = int(os.popen(\"wc -l %s\"%amr).read().split()[0])\n",
    "#\n",
    "con = os.path.join(amr_dir, \"Control.bed\")\n",
    "con_num = int(os.popen(\"wc -l %s\"%con).read().split()[0])\n",
    "print(amr_num, con_num)\n",
    "\n",
    "m6a_dir=\"ASM/corr_m6A/corr_results/plot_profile/AMR_region/extend_bed/\"\n",
    "# m6a_dir=\"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig01/bed/Overlap_exon/\"\n",
    "# m6a_dir=\"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig01/bed/Overlap_utr/\"\n",
    "con_list = glob.glob(\"%s/*_control.bed\" % m6a_dir)\n",
    "\n",
    "for m6a_con in con_list:\n",
    "    m6a = m6a_con.replace(\"_control.bed\", \".bed\")\n",
    "    a=int(os.popen(\"bedtools intersect -a %s -b %s -wa | sort -u | wc -l\" % (amr, m6a)).read().split()[0])\n",
    "    b=int(os.popen(\"bedtools intersect -a %s -b %s -wa | sort -u | wc -l\" % (con, m6a)).read().split()[0])\n",
    "    # m6a_res = os.path.join(result_dir, os.path.basename(m6a))\n",
    "    # os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" % (m6a, amr, m6a_res))\n",
    "    # con_res = os.path.join(result_dir, \"Control-\" + os.path.basename(m6a))\n",
    "    # os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" % (m6a, con, con_res))\n",
    "    c=int(os.popen(\"bedtools intersect -a %s -b %s -wa | sort -u |wc -l\" % (amr,m6a_con)).read().split()[0])\n",
    "    d=int(os.popen(\"bedtools intersect -a %s -b %s -wa | sort -u |wc -l\" % (con,m6a_con)).read().split()[0])\n",
    "    # c,d =  #amr_num-overlap_amr, con_num-overlap_con\n",
    "    oddsratio,pvalue = stats.fisher_exact([[a,b], [c,d]])\n",
    "#     print(\"#############################\")\n",
    "#     print(\"###                AMR   Control\")\n",
    "#     print(\"# m6A              %d       %d  \" % (a,b))\n",
    "#     print(\"# m6A-control      %d       %d  \" % (c,d))\n",
    "#     print(\"#############################\\n\\n\")\n",
    "    print(\"\\t\".join([os.path.basename(m6a), str(pvalue), str(oddsratio), str(a), str(b), str(c), str(d)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRR055537.bed\t0.37334812715935706\t8\t13\n",
      "CRR042308.bed\t0.34829931312955675\t7\t12\n",
      "CRR042318.bed\t0.3349683710848774\t138\t122\n",
      "CRR055539.bed\t0.31354821313944026\t6\t11\n",
      "CRR042296.bed\t0.5263286691753057\t37\t31\n",
      "CRR073018.bed\t0.7251583554189635\t19\t16\n",
      "CRR055533.bed\t0.16103406547895396\t7\t2\n",
      "CRR042280.bed\t0.4739576365096251\t34\t41\n",
      "CRR042310.bed\t0.29563617950838683\t15\t9\n",
      "CRR055541.bed\t0.21103345968256193\t8\t3\n",
      "CRR055543.bed\t1.0\t20\t19\n",
      "brain_IP_1.bed\t0.9037392278095329\t147\t144\n",
      "mus_IP_4.bed\t0.668639306987168\t24\t28\n",
      "CRR055536.bed\t0.11179261492894832\t27\t16\n",
      "sto_IP_5.bed\t0.7866106366767907\t30\t27\n",
      "brain_IP_3.bed\t0.12842241160207574\t120\t145\n",
      "CRR055527.bed\t0.628793213001253\t54\t60\n",
      "CRR042288.bed\t0.11976585674851525\t15\t7\n",
      "CRR042292.bed\t0.05301871827049351\t28\t45\n",
      "CRR042284.bed\t0.7370627390096713\t232\t240\n",
      "CRR042278.bed\t0.2605468972005698\t113\t131\n",
      "CRR055563.bed\t0.27723801745550064\t2\t6\n",
      "CRR055547.bed\t0.6544878500513864\t27\t23\n",
      "sto_IP_4.bed\t0.7665496815715261\t56\t52\n",
      "CRR042298.bed\t0.6333302960591218\t23\t19\n",
      "CRR042320.bed\t1.0\t9\t8\n",
      "liver_IP_1.bed\t0.23545154526445827\t40\t52\n",
      "CRR042290.bed\t0.2191022575641714\t13\t21\n",
      "CRR042302.bed\t0.3094951490133826\t10\t16\n",
      "CRR042314.bed\t0.14516008930541238\t20\t31\n",
      "liver_IP_2.bed\t0.9196083493755289\t53\t51\n",
      "CRR055561.bed\t1.0\t13\t14\n",
      "CRR042294.bed\t0.42268934270290837\t24\t18\n",
      "CRR055553.bed\t0.5391827852915949\t20\t25\n",
      "CRR042312.bed\t0.3508002937256473\t26\t19\n",
      "CRR055549.bed\t0.00712178619026794\t5\t18\n",
      "CRR042282.bed\t0.24805592839489982\t27\t37\n",
      "CRR055531.bed\t0.7374803517816169\t17\t20\n",
      "CRR042316.bed\t1.0\t41\t41\n",
      "lung_IP_5.bed\t0.7961555566263274\t33\t30\n",
      "CRR055559.bed\t1.0\t4\t4\n",
      "pla_IP_4.bed\t0.6862599055841951\t121\t114\n",
      "brain_IP_2.bed\t0.21837630050463092\t72\t88\n",
      "heart_IP_1.bed\t0.6342299504262314\t40\t35\n",
      "CRR073020.bed\t1.0\t4\t5\n",
      "CRR042300.bed\t1.0\t19\t18\n",
      "heart_IP_3.bed\t0.9284049926254774\t67\t65\n",
      "pla_IP_2.bed\t0.15340917216696381\t94\t115\n",
      "CRR055555.bed\t0.10040781457472389\t4\t11\n",
      "pla_IP_6.bed\t0.06662712985335925\t59\t81\n",
      "CRR073016.bed\t0.6049928749007387\t16\t20\n",
      "kid_IP_2.bed\t1.0\t150\t149\n",
      "heart_IP_2.bed\t0.7919343955210076\t71\t67\n",
      "liver_IP_3.bed\t0.5957109052205545\t44\t50\n",
      "mus_IP_5.bed\t0.5717633427619634\t111\t102\n",
      "kid_IP_4.bed\t1.0\t99\t98\n",
      "CRR055525.bed\t0.29185753489087013\t115\t132\n",
      "CRR042306.bed\t0.8815515342869921\t23\t25\n",
      "CRR042304.bed\t0.13666514562109067\t115\t139\n",
      "CRR055529.bed\t0.47942783513860876\t56\t48\n",
      "CRR055557.bed\t0.7665195342988114\t5\t7\n",
      "kid_IP_3.bed\t0.7089664671581364\t92\t98\n",
      "CRR055551.bed\t0.06595276545523418\t12\t4\n",
      "lung_IP_4.bed\t0.5299321885430677\t69\t61\n",
      "CRR055545.bed\t0.6241812643234141\t18\t22\n",
      "CRR042286.bed\t1.0\t39\t38\n"
     ]
    }
   ],
   "source": [
    "### AMR region \n",
    "amr =\"/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/ASM_2kb.bed\"\n",
    "amr_num = int(os.popen(\"wc -l %s\"%amr).read().split()[0])\n",
    "m6a_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/corr_m6A/corr_results/plot_profile/AMR_region/extend_bed/\"\n",
    "con_list = glob.glob(\"%s/*_control.bed\" % m6a_dir)\n",
    "\n",
    "for con in con_list:\n",
    "    m6a = con.replace(\"_control.bed\", \".bed\")\n",
    "#\n",
    "    m6a_num = int(os.popen(\"wc -l %s\"%m6a).read().split()[0])\n",
    "    con_num = int(os.popen(\"wc -l %s\"%con).read().split()[0])\n",
    "    # print(m6a_num, con_num)\n",
    "#\n",
    "    overlap_m6a=int(os.popen(\"bedtools intersect -a %s -b %s -wa | sort -u | wc -l\" % (m6a,amr)).read().split()[0])\n",
    "    overlap_con=int(os.popen(\"bedtools intersect -a %s -b %s -wa | sort -u | wc -l\" % (con,amr)).read().split()[0])\n",
    "    # m6a_res = os.path.join(result_dir, os.path.basename(m6a))\n",
    "    # os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" % (m6a, amr, m6a_res))\n",
    "    # con_res = os.path.join(result_dir, \"Control-\" + os.path.basename(m6a))\n",
    "    # os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" % (m6a, con, con_res))\n",
    "    a,b,c,d = overlap_m6a, overlap_con, m6a_num-overlap_m6a, con_num-overlap_con\n",
    "    oddsratio,pvalue = stats.fisher_exact([[a,b], [c,d]])\n",
    "#     print(\"#############################\")\n",
    "#     print(\"###          m6A   Control\")\n",
    "#     print(\"# AMR         %d       %d  \" % (a,b))\n",
    "#     print(\"# Non-AMR     %d       %d  \" % (c,d))\n",
    "#     print(\"\\t\".join([os.path.basename(m6a), str(pvalue), str(oddsratio)]))\n",
    "#     print(\"#############################\\n\\n\")\n",
    "    print(\"\\t\".join([os.path.basename(m6a), str(pvalue), str(a), str(b)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRR055529_extended.out 0.039 0.039 0.7636\n",
      "kid_IP_2_extended.out 0.001 0.001 0.56\n",
      "CRR042304_extended.out 0.001 0.001 0.6329\n",
      "CRR042292_extended.out 0.016 0.016 0.6305\n",
      "CRR042282_extended.out 0.002 0.002 0.5497\n",
      "brain_IP_1_extended.out 0.022 0.022 0.7119\n",
      "CRR055527_extended.out 0.023 0.023 0.7135\n",
      "CRR073018_extended.out 0.224 0.224 0.8212\n",
      "CRR042286_extended.out 0.005 0.005 0.56\n",
      "CRR055549_extended.out 0.196 0.196 1.2487\n",
      "CRR042312_extended.out 0.375 0.375 0.915\n",
      "CRR042294_extended.out 0.23 0.23 0.821\n",
      "CRR073020_extended.out 0.445 0.445 0.9422\n",
      "CRR055547_extended.out 0.016 0.016 1.565\n",
      "CRR055545_extended.out 0.022 0.022 1.4628\n",
      "brain_IP_3_extended.out 0.001 0.001 0.5672\n",
      "sto_IP_4_extended.out 0.016 0.016 0.4948\n",
      "pla_IP_4_extended.out 0.033 0.033 0.7027\n",
      "lung_IP_4_extended.out 0.003 0.003 0.4581\n",
      "CRR042316_extended.out 0.31 0.31 1.101\n",
      "CRR042298_extended.out 0.164 0.164 0.7802\n",
      "CRR042284_extended.out 0.006 0.006 0.7738\n",
      "CRR073016_extended.out 0.001 0.001 0.4864\n",
      "CRR055525_extended.out 0.005 0.005 0.7207\n",
      "CRR042278_extended.out 0.001 0.001 0.6027\n",
      "heart_IP_3_extended.out 0.153 0.153 0.7805\n",
      "liver_IP_1_extended.out 0.163 0.163 0.7644\n",
      "CRR042310_extended.out 0.027 0.027 0.5391\n",
      "CRR055536_extended.out 0.418 0.418 0.9463\n",
      "kid_IP_3_extended.out 0.025 0.025 0.6292\n",
      "CRR042288_extended.out 0.241 0.241 1.1813\n",
      "CRR055553_extended.out 0.004 0.004 0.528\n",
      "sto_IP_5_extended.out 0.023 0.023 0.4536\n",
      "CRR042320_extended.out 0.334 0.334 0.8453\n",
      "pla_IP_6_extended.out 0.005 0.005 0.506\n",
      "CRR055537_extended.out 0.007 0.007 0.4728\n",
      "CRR055533_extended.out 0.206 0.206 1.2774\n",
      "CRR042280_extended.out 0.408 0.408 0.9475\n",
      "CRR055539_extended.out 0.119 0.119 0.6569\n",
      "CRR055541_extended.out 0.101 0.101 0.5971\n",
      "CRR055555_extended.out 0.407 0.407 1.0513\n",
      "brain_IP_2_extended.out 0.022 0.022 0.6132\n",
      "CRR042306_extended.out 0.176 0.176 0.7978\n",
      "liver_IP_2_extended.out 0.335 0.335 1.085\n",
      "CRR042302_extended.out 0.459 0.459 1.0025\n",
      "CRR055543_extended.out 0.02 0.02 1.6241\n",
      "CRR055561_extended.out 0.156 0.156 0.7332\n",
      "CRR055531_extended.out 0.057 0.057 0.7106\n",
      "CRR042308_extended.out 0.103 0.103 0.6914\n",
      "CRR042290_extended.out 0.072 0.072 0.6834\n",
      "heart_IP_1_extended.out 0.125 0.125 0.6942\n",
      "CRR042314_extended.out 0.175 0.175 0.7971\n",
      "kid_IP_4_extended.out 0.182 0.182 0.8269\n",
      "heart_IP_2_extended.out 0.107 0.107 0.753\n",
      "CRR055557_extended.out 0.093 0.093 0.502\n",
      "CRR055551_extended.out 0.276 0.276 0.8223\n",
      "liver_IP_3_extended.out 0.028 0.028 0.5373\n",
      "CRR055563_extended.out 0.5 0.5 0.972\n",
      "CRR042296_extended.out 0.402 0.402 1.0405\n",
      "CRR055559_extended.out 0.276 0.276 1.2051\n",
      "mus_IP_5_extended.out 0.276 0.276 0.8819\n",
      "mus_IP_4_extended.out 0.003 0.003 0.2717\n",
      "pla_IP_2_extended.out 0.002 0.002 0.5694\n",
      "CRR042300_extended.out 0.133 0.133 0.7514\n",
      "lung_IP_5_extended.out 0.338 0.338 0.8669\n",
      "CRR042318_extended.out 0.004 0.004 0.708\n"
     ]
    }
   ],
   "source": [
    "# GAT enrichment analysis\n",
    "\n",
    "#--annotations=/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig/bed/extend_bed/500bp/{}.bed\n",
    "\n",
    "# ls *.bed | cut -d \".\" -f1 | parallel -j10 \\\n",
    "# \"gat-run.py --ignore-segment-tracks \\\n",
    "# --segments=/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/extend_ASM.bed \\\n",
    "# --annotations=/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/genome_region/basic_anno.bed \\\n",
    "# --workspace=/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/allchrs/merge/AllelicEpigenome-chr1-22_nonoverlap_extend.bed \\\n",
    "# --num-samples=1000 --log=/home/galaxy/project/alleleSpecific_analysis/results/ASM/corr_m6A/corr_results/GAT/{}.log \\\n",
    "# > /home/galaxy/project/alleleSpecific_analysis/results/ASM/corr_m6A/corr_results/GAT/{}.out\"\n",
    "\n",
    "result_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/corr_m6A/corr_results/GAT/\"\n",
    "out_list = glob.glob(\"%s/*.out\"%result_dir)\n",
    "for out in out_list:\n",
    "    df = pd.read_table(out)\n",
    "    print(os.path.basename(out), df['pvalue'][0], df['qvalue'][0], df['fold'][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRR042300.name\t0.19634633517644726\t901\t960\n",
      "CRR055525.name\t0.253894460198616\t1822\t1988\n",
      "CRR042278.name\t0.37625581752693027\t1720\t1890\n",
      "lung_IP_4.name\t0.3644897035809791\t1228\t1340\n",
      "CRR055559.name\t0.900459176078474\t278\t310\n",
      "kid_IP_2.name\t0.43535388780890694\t1459\t1605\n",
      "CRR055545.name\t0.46405768344581694\t853\t931\n",
      "lung_IP_5.name\t0.26126120731514846\t611\t648\n",
      "CRR042280.name\t0.46346493772829633\t1181\t1297\n",
      "heart_IP_2.name\t0.19614268828978026\t832\t884\n",
      "CRR055533.name\t0.616196547334454\t330\t359\n",
      "CRR073018.name\t0.47349797750396\t771\t841\n",
      "CRR042298.name\t0.3576517118877005\t920\t998\n",
      "mus_IP_4.name\t1.0\t450\t508\n",
      "CRR042294.name\t0.2957291322193981\t824\t886\n",
      "CRR073016.name\t0.4579354743899753\t829\t904\n",
      "CRR042304.name\t0.4528530716810406\t2178\t2411\n",
      "pla_IP_4.name\t0.37788438982550615\t1179\t1287\n",
      "sto_IP_5.name\t0.17842784995901828\t569\t595\n",
      "CRR055537.name\t0.25932900420053273\t892\t957\n",
      "brain_IP_1.name\t0.43245935878828357\t1289\t1415\n",
      "brain_IP_3.name\t0.8719280605831534\t1271\t1427\n",
      "CRR042284.name\t0.23715886436770334\t2262\t2476\n",
      "CRR055539.name\t0.45574192172924244\t517\t558\n",
      "mus_IP_5.name\t0.21390815227669807\t1179\t1269\n",
      "heart_IP_1.name\t0.47863722435119377\t526\t569\n",
      "CRR055561.name\t0.5028407526902064\t759\t830\n",
      "CRR042292.name\t0.1473261075649196\t1346\t1443\n",
      "CRR055527.name\t0.46940079824246883\t1372\t1511\n",
      "liver_IP_3.name\t0.8537497767586442\t717\t802\n",
      "CRR055555.name\t0.7358361661321731\t431\t476\n",
      "CRR042310.name\t0.321901040334859\t748\t804\n",
      "CRR042318.name\t0.12506416692416367\t2043\t2209\n",
      "kid_IP_4.name\t0.43056166352608505\t1208\t1324\n",
      "CRR055531.name\t0.4120019765118486\t1109\t1212\n",
      "CRR042312.name\t0.1828572961488173\t782\t828\n",
      "CRR042296.name\t0.2468882651059205\t1088\t1173\n",
      "kid_IP_3.name\t0.5066341758803151\t1047\t1151\n",
      "sto_IP_4.name\t0.5347020430298359\t890\t978\n",
      "CRR042316.name\t0.19755334041757\t1053\t1128\n",
      "heart_IP_3.name\t0.7023081429678604\t772\t856\n",
      "CRR055549.name\t0.6528102214297586\t411\t450\n",
      "CRR055553.name\t0.18039136637143438\t1081\t1156\n",
      "CRR073020.name\t0.8684470738212462\t282\t314\n",
      "CRR042302.name\t0.8316013392482335\t533\t594\n",
      "pla_IP_6.name\t0.4018132253425915\t889\t967\n",
      "CRR042314.name\t0.7017133589100707\t1002\t1114\n",
      "CRR055529.name\t0.25962306382335076\t1567\t1705\n",
      "CRR042308.name\t0.6329796340258953\t711\t784\n",
      "CRR042286.name\t0.12134512417644713\t1424\t1523\n",
      "pla_IP_2.name\t0.49550431200850564\t1216\t1339\n",
      "CRR042290.name\t0.15688326961377763\t963\t1023\n",
      "CRR042320.name\t0.3946886505064017\t625\t674\n",
      "CRR055563.name\t0.7072970120384969\t347\t381\n",
      "CRR055536.name\t0.28680249136907765\t1005\t1085\n",
      "CRR055541.name\t0.4128528557213964\t359\t382\n",
      "CRR042288.name\t0.7589208050313525\t523\t580\n",
      "CRR055543.name\t0.6699139260362565\t615\t679\n",
      "brain_IP_2.name\t0.42048302667132287\t699\t758\n",
      "liver_IP_2.name\t0.6240915712820705\t755\t832\n",
      "CRR055551.name\t0.5871637400541678\t485\t530\n",
      "CRR055557.name\t0.6522856718990586\t348\t380\n",
      "CRR042306.name\t0.1690439520782559\t953\t1014\n",
      "CRR042282.name\t0.17445616372486553\t1458\t1571\n",
      "CRR055547.name\t0.8357499128193708\t742\t829\n",
      "liver_IP_1.name\t0.9798690839912481\t781\t881\n"
     ]
    }
   ],
   "source": [
    "### Gene sets overlap\n",
    "import random\n",
    "from scipy import stats\n",
    "\n",
    "# select control\n",
    "# asm = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/ASM.bed\"\n",
    "# asm_list = pd.read_table(asm, header=None).iloc[:,0]\n",
    "# workspace = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/allchrs/merge/AllelicEpigenome-chr1-22_nonoverlap.bed\"\n",
    "# result_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/control/\"\n",
    "# wk_dict = {}\n",
    "# df = pd.read_table(workspace, header=None)\n",
    "# df.columns = [\"chr\",\"start\",\"end\"]\n",
    "# for a, values in df.iterrows():\n",
    "#     wk_dict[values['chr']] = wk_dict.get(values['chr'], []) + [values.tolist()]\n",
    "\n",
    "# for i in range(1000):\n",
    "#     print(i)\n",
    "#     random_list, result_file = [], os.path.join(result_dir, \"random_%d.bed\"%i)\n",
    "#     for x in asm_list:\n",
    "#         i_list = wk_dict[x]\n",
    "#         random_list.append(random.choice(i_list))\n",
    "#     df_random = pd.DataFrame(random_list)\n",
    "#     df_random.to_csv(result_file, sep=\"\\t\", header=False, index=False)\n",
    "#\n",
    "m6a_dir=\"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig/bed/overlap_withGene/\"\n",
    "m6a_list = glob.glob(\"%s/*.name\"%m6a_dir)\n",
    "#\n",
    "con_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/control/overlap_Gene/\"\n",
    "con_list = glob.glob(\"%s/*.name\"%con_dir)\n",
    "#\n",
    "asm_gene = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/ASM_Gene.name\"\n",
    "asm_number = int(os.popen(\"less %s | wc -l\" % asm_gene).read().split()[0])\n",
    "asm = pd.read_table(asm_gene, header=None).iloc[:,0].tolist()\n",
    "\n",
    "for m6a_gene in m6a_list:\n",
    "    m6a = pd.read_table(m6a_gene, header=None).iloc[:,0].tolist()\n",
    "    overlap_m6a = len([x for x in asm if x in m6a])\n",
    "    a_list, b_list, c_list, d_list = [],[],[],[]\n",
    "    for i in range(100):\n",
    "        con_gene = os.path.join(con_dir, \"random_%d_Gene.name\" % i)\n",
    "        con = pd.read_table(con_gene, header=None).iloc[:,0].tolist()\n",
    "        con_number = len(con)\n",
    "        overlap_con = len([x for x in con if x in m6a])\n",
    "        # pvalue = stats.binom_test(overlap, n=asm_number, p=0.5, alternative='greater')\n",
    "        a,b,c,d = overlap_m6a, overlap_con, asm_number-overlap_m6a, con_number-overlap_con\n",
    "        a_list.append(a)\n",
    "        b_list.append(b)\n",
    "        c_list.append(c)\n",
    "        d_list.append(d)\n",
    "    a,b,c,d = int(np.mean(a_list)), int(np.mean(b_list)),int(np.mean(c_list)),int(np.mean(d_list))\n",
    "    oddsratio,pvalue = stats.fisher_exact([[a,b], [c,d]])\n",
    "    print(\"\\t\".join([os.path.basename(m6a_gene), str(pvalue), str(a), str(b)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# amr overlap exon\n",
    "gtf = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19_exon.bed\"\n",
    "amr_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASM/allelicEpigenome/\"\n",
    "amr = os.path.join(amr_dir, \"extend_ASM.bed\")\n",
    "amr_res = os.path.join(amr_dir, \"extend_exon_ASM.bed\")\n",
    "os.system(\"bedtools intersect -a %s -b %s -wa | sort -u > %s\" % (amr, gtf, amr_res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# m6A overlap cds or exon\n",
    "gtf = \"/home/galaxy/project/alleleSpecific_analysis/data/hg19_genome/hg19_utr.bed\"\n",
    "m6a_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig01/bed/\"\n",
    "m6a_list = glob.glob(\"%s/*.bed\" % m6a_dir)\n",
    "result_dir = os.path.join(m6a_dir, \"Overlap_utr\")\n",
    "os.system(\"mkdir -p %s\" % result_dir)\n",
    "for m6a in m6a_list:\n",
    "    res = os.path.join(result_dir, os.path.basename(m6a))\n",
    "    os.system(\"bedtools intersect -a %s -b %s -wa | sort -u > %s\" % (m6a, gtf, res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "## extend arrounding 1000bp\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig01/bed/\")\n",
    "result_dir = \"extended_1000bp/\"\n",
    "bed_list = glob.glob(\"*.bed\")\n",
    "for bed in bed_list:\n",
    "    res = os.path.join(result_dir, bed)\n",
    "    df = pd.read_table(bed, header=None)\n",
    "    # df.columns = ['chr','s','e']\n",
    "    df['s'] = df.iloc[:,1] - 1000\n",
    "    df['e'] = df.iloc[:,1] + 1001\n",
    "    df['chr'] = df.iloc[:,0]\n",
    "    df = df[['chr','s','e']].drop_duplicates()\n",
    "    df.to_csv(res, sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
