{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import datetime\n",
    "import shutil\n",
    "import random\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import subprocess\n",
    "import matplotlib.mlab as mlab\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'svg'\n",
    "from multiprocessing import Pool\n",
    "from sklearn.utils import shuffle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## rbp_dir = \"/data4/galaxy/project/m6a_R-loop_snp/RBP/combine_CLIPdb-starBase2_Encode\"\n",
    "# In the m6a paper, the rbp_dir was used. However, in the study, we need to get the tissue information.\n",
    "# Based on the above reason, we decided to use the data from the CLIPDB and Starbase2 database, where we can get the tissue  info!\n",
    "#CLIPdb database didn't contain tissue data, current now!\n",
    "## 2019年1月29日\n",
    "#The CLIPdb database was renamed to POSTAR.\n",
    "# retrieve the newest data from POSTAR & Starbase2 by python spider script."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### POSTAR Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Total RBP = 2019_1_29 + second_result = 171\n",
    "# rbp_dir = \"/home/galaxy/data/RBP/POSTAR/2019_1_29/\"\n",
    "# rbp_dir = \"/home/lzg/postar/second_result/\"\n",
    "rbp_dir = \"/home/lzg/trush/gene_overlap/RBP/second_result/\"\n",
    "tmp_dir = \"/home/galaxy/data/RBP/POSTAR/tmp\"\n",
    "if not os.path.exists(tmp_dir):\n",
    "    os.makedirs(tmp_dir)\n",
    "result_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/info_dir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAPRIN1\n",
      "SF3B1\n",
      "FIP1L1\n",
      "DDX59\n",
      "YTHDF1\n",
      "EIF3G\n",
      "IGF2BP3\n",
      "SUPV3L1\n",
      "FXR1\n",
      "UPF1\n",
      "ATXN2\n",
      "STAU1\n",
      "METAP2\n",
      "ZC3H7B\n",
      "EFTUD2\n",
      "NOL12\n",
      "YTHDC1\n",
      "RTCB\n",
      "CELF2\n",
      "RPS5\n",
      "EWSR1\n",
      "EIF3A\n",
      "YBX3\n",
      "EIF3H\n",
      "CPSF2\n",
      "EIF3D\n",
      "DDX3X\n",
      "GRWD1\n",
      "GEMIN5\n",
      "YTHDF3\n",
      "EIF3B\n",
      "RBM15B\n",
      "NUDT21\n",
      "SUGP2\n",
      "CSTF2T\n",
      "NONO\n",
      "FXR2\n",
      "YTHDF2\n",
      "YTHDC2\n",
      "TNRC6A\n",
      "RBFOX2\n",
      "CPSF7\n",
      "PCBP2\n",
      "LIN28A\n",
      "CPSF6\n",
      "HNRNPU\n",
      "SF3B4\n",
      "RBM15\n",
      "TBRG4\n",
      "SUB1\n",
      "QKI\n",
      "IGF2BP1\n",
      "FUS\n",
      "METTL14\n",
      "TIAL1\n",
      "BUD13\n",
      "PPIG\n",
      "UCHL5\n",
      "HNRNPH1\n",
      "ILF3\n",
      "DGCR8\n",
      "XPO5\n",
      "MTPAP\n",
      "IGF2BP2\n",
      "RBM22\n",
      "ALKBH5\n",
      "RBM10\n",
      "HNRNPM\n",
      "WDR33\n",
      "KHSRP\n",
      "DDX55\n",
      "NOP58\n",
      "EIF4A3\n",
      "GPKOW\n",
      "NCBP3\n",
      "PRKRA\n",
      "CPSF1\n",
      "CPSF4\n",
      "GTF2F1\n",
      "SSB\n",
      "MBNL2\n",
      "KHDRBS1\n",
      "PUM2\n",
      "LIN28B\n",
      "SRSF1\n",
      "WTAP\n",
      "FBL\n",
      "EIF4G2\n",
      "GRSF1\n",
      "NCBP2\n",
      "TARBP2\n",
      "DICER1\n",
      "U2AF1\n",
      "CPSF3\n",
      "FMR1\n",
      "SAFB2\n",
      "SRRM4\n",
      "TIA1\n",
      "FTO\n",
      "RBM47\n",
      "CDC40\n",
      "HNRNPK\n",
      "RBM27\n",
      "SRSF7\n",
      "FUBP3\n",
      "HNRNPD\n",
      "SLTM\n",
      "LSM11\n",
      "ZNF622\n",
      "DROSHA\n",
      "SF3A3\n",
      "XRCC6\n",
      "DDX6\n",
      "TAF15\n",
      "AGGF1\n",
      "AARS\n",
      "FAM120A\n",
      "DHX30\n",
      "HNRNPF\n",
      "DKC1\n",
      "DDX24\n",
      "DDX42\n",
      "NOP56\n",
      "CNBP\n",
      "MOV10\n",
      "PRPF8\n",
      "RPS3\n",
      "AKAP8L\n",
      "FASTKD2\n",
      "XRN2\n",
      "TNRC6C\n",
      "TROVE2\n",
      "METTL3\n",
      "SMNDC1\n",
      "LARP4\n",
      "RBPMS\n",
      "SRSF9\n",
      "TRA2A\n",
      "ZRANB2\n",
      "ALKBH1\n",
      "SERBP1\n",
      "PUS1\n",
      "NKRF\n",
      "SND1\n",
      "FKBP4\n",
      "LARP7\n",
      "DIS3L2\n",
      "RBM5\n",
      "GNL3\n",
      "BCCIP\n",
      "YWHAG\n",
      "HNRNPA2B1\n",
      "AUH\n",
      "NPM1\n",
      "HNRNPUL1\n",
      "SFPQ\n",
      "RPS11\n",
      "HLTF\n",
      "NSUN2\n",
      "TNRC6B\n",
      "PPIL4\n",
      "EZH2\n",
      "SLBP\n",
      "SBDS\n"
     ]
    }
   ],
   "source": [
    "def get_dict():\n",
    "    rbp_list = glob.glob(\"%s/*.txt\" % rbp_dir)\n",
    "    rbp_dict = {}\n",
    "    for rbp in rbp_list:\n",
    "        rbp_name = os.path.basename(rbp).split(\"_\")[0]\n",
    "        rbp_dict[rbp_name] = rbp_dict.get(rbp_name, []) + [rbp]\n",
    "    return rbp_dict\n",
    "\n",
    "def process_one_targetGene(rbp):\n",
    "    result_file = os.path.join(tmp_dir, os.path.basename(rbp))\n",
    "    target = os.path.basename(rbp).split(\".txt\")[0].split(\"_\")[1]\n",
    "    df = pd.read_table(rbp, sep=\"\\t\", header=None)\n",
    "    df.columns = [\"score1\", \"score2\", \"Gene ID\", \"Gene Name\", \"Gene Type\", \"Genomic context\", \"Position\", \"CLIP-seq technology and peak calling method\", \"db\", \".\", \"Strand\", \"Tissue type\", \"Transcript ID\"]\n",
    "    selected_col = [\"Gene ID\", \"Tissue type\", \"db\", \"Position\"]\n",
    "    df = df[selected_col].drop_duplicates()\n",
    "    df[\"name\"] = df[\"Gene ID\"] + \";\" + df[\"db\"] + \";\" + df[\"Position\"]\n",
    "    df[\"Chr name\"] = df[\"Position\"].str.split(\":\").str[0]\n",
    "    df[\"Start\"] = df[\"Position\"].str.split(\":\").str[1].str.split(\"-\").str[0]\n",
    "    df[\"End\"] = df[\"Position\"].str.split(\"-\").str[1]\n",
    "    df_result = df[[\"Gene ID\", \"Tissue type\", \"Chr name\", \"Start\", \"End\", \"name\"]]\n",
    "    df_result.to_csv(result_file, sep=\"\\t\", header=False, index=False)\n",
    "    \n",
    "def process_one_rbp(rbp_name, rbp_list):\n",
    "    print(rbp_name)\n",
    "    result_total_file = os.path.join(result_dir, \"%s.txt\" % rbp_name)\n",
    "    if (not os.path.exists(result_total_file)) and len(glob.glob(\"%s/*\" % tmp_dir)) == 0:\n",
    "#     if len(glob.glob(\"%s/*\" % tmp_dir)) == 0:\n",
    "        pool = Pool(processes=30)\n",
    "        for rbp in rbp_list:\n",
    "            pool.apply_async(process_one_targetGene, (rbp, ))\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "        os.system(\"cat %s/*.txt > %s\" % (tmp_dir, result_total_file))\n",
    "        os.system(\"rm %s/*\" % tmp_dir)\n",
    "    else:\n",
    "        print(\"tmp dir exits file!\")\n",
    "        \n",
    "def main_method():\n",
    "    rbp_dict = get_dict()\n",
    "    for rbp_name, rbp_list in rbp_dict.items():\n",
    "        if len(rbp_list) > 0:\n",
    "            process_one_rbp(rbp_name, rbp_list)\n",
    "        else:\n",
    "            print(rbp_name)\n",
    "    os.system(\"rm -rf %s\" % tmp_dir)\n",
    "#################\n",
    "main_method()\n",
    "#################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Relation analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Ignore tissue information\n",
    "#### format txt into bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/info_dir/\"\n",
    "result_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/GRCh38_bed_dir/\"\n",
    "txt_list = glob.glob(\"%s/*.txt\" % data_dir)\n",
    "for txt in txt_list:\n",
    "    result_bed = os.path.join(result_dir, os.path.basename(txt).replace(\".txt\", \".bed\"))\n",
    "    os.system(\"cut -f 3-6 %s | sort -k1,1 -k2,2n > %s\" % (txt, result_bed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/GRCh38_bed_dir\"\n",
    "result_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/GRCh38_merged_bed\"\n",
    "raw_beds = glob.glob(\"%s/*.bed\" % data_dir)\n",
    "for raw_bed in raw_beds:\n",
    "    result_bed = os.path.join(result_dir, os.path.basename(raw_bed))\n",
    "    os.system(\"bedtools merge -i %s > %s\" % (raw_bed, result_bed))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### generate Data Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################################################\n",
    "rbp_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/GRCh38_bed_dir/\"\n",
    "eQTL_dir = \"/home/galaxy/data/QTL/m6a_related_SNPs/eQTL/m6a_qtl_bed/\"\n",
    "result_dir = \"/home/galaxy/project/QTL_analysis/related_with_RBP/POSTAR/data_table\"\n",
    "######################################################################################\n",
    "rbp_list = glob.glob(\"%s/*.bed\" % rbp_dir)\n",
    "eqtl_list = glob.glob(\"%s/*.bed\" % eQTL_dir)\n",
    "for eqtl in eqtl_list:\n",
    "    for rbp in rbp_list:\n",
    "        basename = \"%s-%s\" % (os.path.basename(eqtl).split(\".bed\")[0], os.path.basename(rbp))\n",
    "        result_file = os.path.join(result_dir, basename)\n",
    "        os.system(\"bedtools intersect -a %s -b %s -wa -wb > %s\" % (eqtl, rbp, result_file))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "###\n",
    "rbp_dir = \"/home/galaxy/data/RBP/POSTAR/POSTAR_bed/GRCh38_bed_dir/\"\n",
    "###\n",
    "eQTL_dir = \"/home/galaxy/data/QTL/m6a_related_SNPs/eQTL/m6a_qtl_bed/\"\n",
    "HapMap_snp = \"/home/galaxy/data/QTL/HapMap_SNP/autosomal_GRCh38_snp.bed\"\n",
    "result_dir = \"/home/galaxy/project/QTL_analysis/related_with_RBP/Fisher_test_RBP\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_list = glob.glob(\"%s/*.bed\" % rbp_dir)\n",
    "eQTL_list = glob.glob(\"%s/*.bed\" % eQTL_dir)\n",
    "HapMap_NUMBER = 81039058"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "for eqtl in eQTL_list:\n",
    "    res_list, pool = [], Pool(processes=30)\n",
    "    for rbp in rbp_list:\n",
    "#         res_list = fisher_test(eqtl, rbp)\n",
    "        res = pool.apply_async(fisher_test, (eqtl, rbp, ))\n",
    "        res_list.append(res)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "    write_to_file(res_list, os.path.join(result_dir, os.path.basename(eqtl).split(\"-\")[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_to_file(res_list, res_file):\n",
    "    res_list = [\"%s\\n\" % x.get() for x in res_list]\n",
    "    with open(res_file, \"w\") as fw:\n",
    "        fw.writelines(res_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fisher_test(eqtl_bed, RPB_bed):\n",
    "         #              RPB_bed       x\n",
    "        # QTL_SNP         a       m6a - a\n",
    "        # HapMap_SNP      c       con - c\n",
    "    QTL_num = stat_file_line_numbers(eqtl_bed)\n",
    "    a = os.popen(\"bedtools intersect -a %s -b %s -wa |sort| uniq | wc -l\" % (eqtl_bed, RPB_bed)).read()\n",
    "    c = os.popen(\"bedtools intersect -a %s -b %s -wa |sort| uniq | wc -l\" % (HapMap_snp, RPB_bed)).read()\n",
    "    b, d = QTL_num - int(a), HapMap_NUMBER - int(c)\n",
    "    oddsratio, pvalue = stats.fisher_exact([[int(a), b], [int(c), d]])\n",
    "    line=\"%s\\t%s\\t%f\\t%f\"%(os.path.basename(eqtl_bed).split(\"-\")[0], os.path.basename(RPB_bed), oddsratio, pvalue)\n",
    "    return line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stat_file_line_numbers(in_file):\n",
    "    raw_return = os.popen(\"wc -l %s\" % in_file)\n",
    "    line_num = int(raw_return.read().split()[0])\n",
    "    return line_num"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
