{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import datetime\n",
    "import shutil\n",
    "import random\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import subprocess\n",
    "from multiprocessing import Pool\n",
    "from statsmodels.stats import multitest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## haploview = \"/home/galaxy/software/Haploview.jar\"\n",
    "# Perform LD analysis for each GWAS SNP to obtain its LD mutations with a parameter of r2 > 0.8\n",
    "# http://icsnpathway.psych.ac.cn/  webserver\n",
    "data_dir = \"/home/galaxy/data/disease/GWAS/ICSNPathway\"\n",
    "result_dir = \"/home/galaxy/data/disease/GWAS/ICSNPathway/result_dir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_list = glob.glob(\"%s/*.txt\" % data_dir)\n",
    "result_file = os.path.join(result_dir, \"GWAS_and_LD-mutations.txt\")\n",
    "df_list = []\n",
    "for data in data_list:\n",
    "    df = pd.read_table(data, sep=\"\\t\")\n",
    "    df[\"query SNP\"] = df[\"Candidate causal SNP\"].str.split().str[0]\n",
    "    df[\"LD SNP\"] = df[\"In LD with\"].str.split().str[0]\n",
    "    df[\"Type\"] = np.where(df[\"query SNP\"]==df[\"LD SNP\"], \"TagSNP\", \"LD_derived\")\n",
    "    df_list.append(df[[\"query SNP\", \"LD SNP\", \"Type\"]])\n",
    "df_result = pd.concat(df_list).drop_duplicates()\n",
    "df_result.to_csv(result_file, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge icsnpathway result with GWAS SNPs\n",
    "icsn_file = \"/home/galaxy/data/disease/GWAS/ICSNPathway/result_dir/GWAS_and_LD-mutations.txt\"\n",
    "gwas = \"/home/galaxy/data/disease/GWAS/total_result.txt\"\n",
    "result_file = \"/home/galaxy/data/disease/GWAS/total_result_with_LD.txt\"\n",
    "df_icsn = pd.read_table(icsn_file, sep=\"\\t\")\n",
    "df_gwas = pd.read_table(gwas, sep=\"\\t\")\n",
    "df_merge = df_gwas.merge(df_icsn, how=\"left\", left_on=\"SNPS\", right_on=\"query SNP\").dropna(how=\"any\")\n",
    "df_merge.to_csv(result_file, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map m6a-related eQTL SNPs to the icsnpathway result\n",
    "##############################################################################################\n",
    "eqtl_dir = \"/home/galaxy/data/QTL/m6a_related_SNPs/eQTL/m6a_qtl_bed/\"\n",
    "gwas_file = \"/home/galaxy/data/disease/GWAS/total_result_with_LD.txt\"\n",
    "result_dir = \"/home/galaxy/project/QTL_analysis/related_with_disease/GWAS\"\n",
    "##############################################################################################\n",
    "df_gwas = pd.read_table(gwas_file, sep=\"\\t\")\n",
    "eqtl_list = glob.glob(\"%s/*.bed\" % eqtl_dir)\n",
    "for eqtl in eqtl_list:\n",
    "    result_file = os.path.join(result_dir, os.path.basename(eqtl).replace(\".bed\", \".txt\"))\n",
    "    df_eqtl = pd.read_table(eqtl, sep=\"\\t\", header=None, names=[\"Chr\", \"Start\", \"End\", \"Info\"])\n",
    "    df_eqtl[\"SNPS\"] = df_eqtl.iloc[:, 3].str.split(\";\").str[2]\n",
    "    df_eqtl[\"Gene\"] = df_eqtl.iloc[:, 3].str.split(\";\").str[0]\n",
    "#     print(df_eqtl.head())\n",
    "    df_merge = df_eqtl.merge(df_gwas, how=\"left\")\n",
    "    df_merge.to_csv(result_file, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_dir = \"/home/galaxy/data/QTL/m6a_related_SNPs/eQTL/m6a_qtl_bed/\"\n",
    "gwas_dir = \"/data5/galaxy/project/GWAS_db/04_class_snp/cluster_by_map/snp_classification/\"\n",
    "result_dir = \"/home/galaxy/project/QTL_analysis/related_with_disease/GWAS/enrichment_result\"\n",
    "HapMap_snp = \"/home/galaxy/data/QTL/HapMap_SNP/autosomal_GRCh38_snp.bed\"\n",
    "hapmap_list = pd.read_table(HapMap_snp, sep=\"\\t\", header=None).iloc[:, -1].tolist()\n",
    "HapMap_NUMBER = 81039058"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "HapMap_NUMBER = 81039058\n",
    "hapset = set(hapmap_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 30\n",
      "0 100\n",
      "0 268\n",
      "1 239\n",
      "0 334\n",
      "0 247\n",
      "1 370\n",
      "0 73\n",
      "0 312\n",
      "3 439\n",
      "0 547\n",
      "2 654\n",
      "0 904\n",
      "2 898\n",
      "3 952\n",
      "3 2431\n",
      "30 1100\n",
      "2 1225\n",
      "0 100\n",
      "0 268\n",
      "18 898\n",
      "4 1225\n",
      "0 30\n",
      "0 73\n",
      "3 239\n",
      "0 247\n",
      "14 439\n",
      "6 334\n",
      "0 312\n",
      "1 547\n",
      "4 370\n",
      "2 904\n",
      "5 952\n",
      "4 654\n",
      "7 1100\n",
      "14 2431\n",
      "0 30\n",
      "0 247\n",
      "0 268\n",
      "0 239\n",
      "0 370\n",
      "1 654\n",
      "0 547\n",
      "0 904\n",
      "0 952\n",
      "0 100\n",
      "0 334\n",
      "2 439\n",
      "0 312\n",
      "17 1100\n",
      "0 1225\n",
      "0 73\n",
      "1 898\n",
      "0 2431\n",
      "0 312\n",
      "0 370\n",
      "1 898\n",
      "21 1100\n",
      "1 654\n",
      "0 30\n",
      "0 904\n",
      "0 73\n",
      "0 100\n",
      "0 247\n",
      "0 239\n",
      "0 268\n",
      "0 334\n",
      "2 439\n",
      "0 547\n",
      "0 952\n",
      "0 1225\n",
      "1 2431\n",
      "0 30\n",
      "0 100\n",
      "0 73\n",
      "0 247\n",
      "1 312\n",
      "4 334\n",
      "10 439\n",
      "0 268\n",
      "3 239\n",
      "1 547\n",
      "4 370\n",
      "4 654\n",
      "1 904\n",
      "7 952\n",
      "4 1100\n",
      "15 898\n",
      "8 1225\n",
      "7 2431\n",
      "0 334\n",
      "0 30\n",
      "0 100\n",
      "0 312\n",
      "0 239\n",
      "0 904\n",
      "0 247\n",
      "1 952\n",
      "0 268\n",
      "0 73\n",
      "2 439\n",
      "2 898\n",
      "0 370\n",
      "0 547\n",
      "23 1100\n",
      "0 1225\n",
      "2 654\n",
      "2 2431\n",
      "0 100\n",
      "0 73\n",
      "5 370\n",
      "0 547\n",
      "9 654\n",
      "7 904\n",
      "0 30\n",
      "0 247\n",
      "0 268\n",
      "3 312\n",
      "2 334\n",
      "2 239\n",
      "8 439\n",
      "11 952\n",
      "19 898\n",
      "13 1100\n",
      "3 1225\n",
      "17 2431\n",
      "0 100\n",
      "1 239\n",
      "0 30\n",
      "0 247\n",
      "0 268\n",
      "0 73\n",
      "3 439\n",
      "0 547\n",
      "0 334\n",
      "3 370\n",
      "0 312\n",
      "4 952\n",
      "5 898\n",
      "3 654\n",
      "0 904\n",
      "26 1100\n",
      "1 1225\n",
      "6 2431\n",
      "0 30\n",
      "0 73\n",
      "0 100\n",
      "1 268\n",
      "1 239\n",
      "0 247\n",
      "0 312\n",
      "0 334\n",
      "3 439\n",
      "0 370\n",
      "1 547\n",
      "1 654\n",
      "0 952\n",
      "4 898\n",
      "3 904\n",
      "23 1100\n",
      "1 1225\n",
      "3 2431\n",
      "0 247\n",
      "0 30\n",
      "0 100\n",
      "2 239\n",
      "2 268\n",
      "0 73\n",
      "3 334\n",
      "0 312\n",
      "4 370\n",
      "9 439\n",
      "3 547\n",
      "4 904\n",
      "7 654\n",
      "47 1100\n",
      "9 898\n",
      "6 1225\n",
      "7 952\n",
      "16 2431\n",
      "4 239\n",
      "4 268\n",
      "7 547\n",
      "0 30\n",
      "0 100\n",
      "0 73\n",
      "1 247\n",
      "9 370\n",
      "0 312\n",
      "2 334\n",
      "10 439\n",
      "7 654\n",
      "6 1225\n",
      "22 898\n",
      "9 952\n",
      "7 904\n",
      "19 1100\n",
      "27 2431\n",
      "1 334\n",
      "0 30\n",
      "0 73\n",
      "0 100\n",
      "0 247\n",
      "1 268\n",
      "0 312\n",
      "3 370\n",
      "2 239\n",
      "5 439\n",
      "1 547\n",
      "4 904\n",
      "7 654\n",
      "7 952\n",
      "39 1100\n",
      "1 1225\n",
      "8 898\n",
      "9 2431\n",
      "0 334\n",
      "0 239\n",
      "0 268\n",
      "3 904\n",
      "0 30\n",
      "0 100\n",
      "0 73\n",
      "0 247\n",
      "1 312\n",
      "4 370\n",
      "3 439\n",
      "0 547\n",
      "3 654\n",
      "5 898\n",
      "0 1225\n",
      "4 952\n",
      "23 1100\n",
      "1 2431\n",
      "0 30\n",
      "1 268\n",
      "1 312\n",
      "4 1100\n",
      "0 100\n",
      "1 239\n",
      "0 73\n",
      "0 247\n",
      "5 370\n",
      "2 547\n",
      "0 334\n",
      "7 439\n",
      "3 904\n",
      "3 654\n",
      "8 898\n",
      "7 1225\n",
      "4 952\n",
      "13 2431\n",
      "0 30\n",
      "0 100\n",
      "0 73\n",
      "2 239\n",
      "0 268\n",
      "0 247\n",
      "0 312\n",
      "2 370\n",
      "3 439\n",
      "0 547\n",
      "1 334\n",
      "2 654\n",
      "7 898\n",
      "0 952\n",
      "25 1100\n",
      "2 904\n",
      "1 1225\n",
      "6 2431\n",
      "0 30\n",
      "0 100\n",
      "0 247\n",
      "2 439\n",
      "0 73\n",
      "0 312\n",
      "0 268\n",
      "0 334\n",
      "0 1225\n",
      "0 370\n",
      "0 239\n",
      "0 547\n",
      "0 654\n",
      "23 1100\n",
      "0 952\n",
      "0 904\n",
      "2 898\n",
      "1 2431\n",
      "0 268\n",
      "0 30\n",
      "0 247\n",
      "10 439\n",
      "2 370\n",
      "0 73\n",
      "0 100\n",
      "0 547\n",
      "1 334\n",
      "0 312\n",
      "4 952\n",
      "1 904\n",
      "2 1225\n",
      "1 239\n",
      "7 898\n",
      "4 654\n",
      "33 1100\n",
      "11 2431\n",
      "0 30\n",
      "0 73\n",
      "0 100\n",
      "2 268\n",
      "4 239\n",
      "7 439\n",
      "2 312\n",
      "0 247\n",
      "2 370\n",
      "0 654\n",
      "2 547\n",
      "0 334\n",
      "2 904\n",
      "3 898\n",
      "2 1100\n",
      "2 1225\n",
      "5 952\n",
      "3 2431\n",
      "0 30\n",
      "0 268\n",
      "0 547\n",
      "0 312\n",
      "0 73\n",
      "5 952\n",
      "0 100\n",
      "25 1100\n",
      "1 239\n",
      "4 654\n",
      "8 898\n",
      "0 247\n",
      "2 370\n",
      "5 439\n",
      "0 334\n",
      "2 904\n",
      "3 1225\n",
      "8 2431\n",
      "0 30\n",
      "0 312\n",
      "0 247\n",
      "0 952\n",
      "0 904\n",
      "0 1225\n",
      "0 73\n",
      "0 100\n",
      "0 334\n",
      "0 547\n",
      "0 654\n",
      "20 1100\n",
      "1 898\n",
      "0 268\n",
      "0 239\n",
      "0 370\n",
      "1 439\n",
      "0 2431\n",
      "0 547\n",
      "5 370\n",
      "0 30\n",
      "0 73\n",
      "0 100\n",
      "0 268\n",
      "13 1100\n",
      "0 247\n",
      "14 439\n",
      "3 312\n",
      "3 239\n",
      "12 654\n",
      "5 334\n",
      "11 898\n",
      "7 952\n",
      "5 904\n",
      "30 2431\n",
      "12 1225\n"
     ]
    }
   ],
   "source": [
    "gwas_list = glob.glob(\"%s/*.bed\" % gwas_dir)\n",
    "eqtl_list = glob.glob(\"%s/*.bed\" % eqtl_dir)\n",
    "for eqtl in eqtl_list:\n",
    "    res_list, pool = [], Pool(processes=25)\n",
    "    for gwas in gwas_list:\n",
    "#         fisher_test(eqtl, gwas)\n",
    "        res = pool.apply_async(fisher_test, (eqtl, gwas, ))\n",
    "        res_list.append(res)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "    write_to_file(res_list, os.path.join(result_dir, os.path.basename(eqtl).split(\"-\")[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stat_file_line_numbers(in_file):\n",
    "    raw_return = os.popen(\"wc -l %s\" % in_file)\n",
    "    line_num = int(raw_return.read().split()[0])\n",
    "    return line_num\n",
    "\n",
    "def fisher_test(eqtl_bed, gwas_bed):\n",
    "         #              gwas_bed       x\n",
    "        # QTL_SNP         a       QTL_SNP - a\n",
    "        # HapMap_SNP      c       HapMap_SNP - c\n",
    "    qtl_list = pd.read_table(eqtl_bed, sep=\"\\t\", header=None).iloc[:, -1].str.split(\";\").str[2].tolist()\n",
    "    gw_list = pd.read_table(gwas_bed, sep=\"\\t\", header=None).iloc[:, -1].tolist()\n",
    "#     print(hapmap_list[0])\n",
    "#     print(qtl_list[0])\n",
    "#     print(gw_list[0])\n",
    "    QTL_num = stat_file_line_numbers(eqtl_bed)\n",
    "    a = len([x for x in qtl_list if x in gw_list])\n",
    "    c = len(hapset.intersection(set(gw_list)))\n",
    "    b, d = QTL_num - int(a), HapMap_NUMBER - int(c)\n",
    "    print(a, c)\n",
    "#     sys.exit(1)\n",
    "#     oddsratio, pvalue = stats.fisher_exact([[int(a), b], [int(c), d]])\n",
    "    oddsratio, pvalue = stats.fisher_exact([[int(a), b], [int(c), d]], alternative=\"less\")\n",
    "    pvalue = 1 - pvalue\n",
    "    line=\"%s\\t%s\\t%f\\t%f\\t%d\"%(os.path.basename(eqtl_bed).split(\"-\")[0],os.path.basename(gwas_bed),oddsratio,pvalue,a)\n",
    "    return line\n",
    "\n",
    "def write_to_file(res_list, res_file):\n",
    "    res_list = [\"%s\\n\" % x.get() for x in res_list]\n",
    "    with open(res_file, \"w\") as fw:\n",
    "        fw.writelines(res_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FDR correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fdr_correction():\n",
    "    result_dir = \"/home/galaxy/project/QTL_analysis/related_with_disease/GWAS/FDR_correction\"\n",
    "    data_dir = \"/home/galaxy/project/QTL_analysis/related_with_disease/GWAS/enrichment_result\"\n",
    "    file_list = glob.glob(\"%s/*.bed\" % data_dir)\n",
    "    for x in file_list:\n",
    "        res_file = os.path.join(result_dir, os.path.basename(x).replace(\".bed\", \".txt\"))\n",
    "        df = pd.read_table(x, sep=\"\\t\", header=None, names=[\"tissue\", \"disease\", \"oddsratio\", \"pvalue\", \"BothPos\"])\n",
    "        pvalue_list = [float(x) for x in df.iloc[:, -2].tolist()]\n",
    "        qvalue_list = bh_correction(pvalue_list)\n",
    "        df[\"qvalue\"] = qvalue_list\n",
    "        df.to_csv(res_file, sep=\"\\t\", index=False)\n",
    "fdr_correction()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bh_correction(pvalue_list):\n",
    "    true_or_false, qvalue_list = multitest.fdrcorrection(pvalue_list, method=\"indep\")\n",
    "    return qvalue_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
