{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import glob\n",
    "import datetime\n",
    "import shutil\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "pd.options.mode.chained_assignment = None  # default='warn'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Chr  Start    End Strand                        Transcript_id\n",
      "2  chr1  11869  12227      +  ENSG00000223972.5;ENST00000456328.2\n",
      "3  chr1  12613  12721      +  ENSG00000223972.5;ENST00000456328.2\n",
      "4  chr1  13221  14409      +  ENSG00000223972.5;ENST00000456328.2\n",
      "6  chr1  12010  12057      +  ENSG00000223972.5;ENST00000450305.2\n",
      "7  chr1  12179  12227      +  ENSG00000223972.5;ENST00000450305.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/galaxy/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n"
     ]
    }
   ],
   "source": [
    "# ~2min\n",
    "human_gff = \"/data/database/GRCh38/GENCODE/gencode.v27.annotation.gff3\"\n",
    "# human_gff = \"/home/galaxy/data/splicing_sites/test.gff3\"\n",
    "names = [\"Chr\", \"DB\", \"Feature\", \"Start\", \"End\", \"Dot_1\", \"Strand\", \"Dot_2\", \"Info\"]\n",
    "df = pd.read_table(human_gff, sep=\"\\t\", header=None, names=names, comment=\"#\")\n",
    "# print(df.head())\n",
    "df[\"Gene_id\"] = df[\"Info\"].str.split(\"gene_id=\").str[1].str.split(\";\").str[0]\n",
    "df[\"Transcript_id\"] = df[\"Info\"].str.split(\"transcript_id=\").str[1].str.split(\";\").str[0]\n",
    "df_exon = df[df[\"Feature\"] == \"exon\"]\n",
    "# print(df_exon.head())\n",
    "df_exon[\"Transcript_id\"] = df_exon[\"Gene_id\"] + \";\" + df_exon[\"Transcript_id\"]\n",
    "df_bed = df_exon[[\"Chr\", \"Start\", \"End\", \"Strand\", \"Transcript_id\"]]\n",
    "print(df_bed.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         Chr      Start        End Strand  \\\n",
      "Transcript_id                                                               \n",
      "ENSG00000000003.14;ENST00000373020.8 0  chrX  100636605  100636607      -   \n",
      "                                     1  chrX  100635746  100635748      -   \n",
      "                                     2  chrX  100635555  100635557      -   \n",
      "                                     3  chrX  100635252  100635254      -   \n",
      "                                     4  chrX  100635175  100635177      -   \n",
      "\n",
      "                                                               Transcript_id  \\\n",
      "Transcript_id                                                                  \n",
      "ENSG00000000003.14;ENST00000373020.8 0  ENSG00000000003.14;ENST00000373020.8   \n",
      "                                     1  ENSG00000000003.14;ENST00000373020.8   \n",
      "                                     2  ENSG00000000003.14;ENST00000373020.8   \n",
      "                                     3  ENSG00000000003.14;ENST00000373020.8   \n",
      "                                     4  ENSG00000000003.14;ENST00000373020.8   \n",
      "\n",
      "                                       Event  \n",
      "Transcript_id                                 \n",
      "ENSG00000000003.14;ENST00000373020.8 0     3  \n",
      "                                     1     5  \n",
      "                                     2     3  \n",
      "                                     3     5  \n",
      "                                     4     3  \n"
     ]
    }
   ],
   "source": [
    "# run time:~20min\n",
    "df_result = df_bed.groupby([\"Transcript_id\"]).apply(get_splice_site)\n",
    "print(df_result.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "# attention! gff3 is one-based, while bed is zero-based format.\n",
    "def get_splice_site(df_i):\n",
    "    splice_list = []\n",
    "    df_i = df_i.drop_duplicates()\n",
    "    if len(df_i) > 1:\n",
    "        # first exon\n",
    "        exon_1 = df_i.iloc[0, :]\n",
    "        res = []\n",
    "        if exon_1[\"Strand\"] == \"+\":\n",
    "            res = [exon_1[\"Chr\"],exon_1[\"End\"],exon_1[\"End\"]+2,exon_1[\"Strand\"],exon_1[\"Transcript_id\"],\"5\"]\n",
    "        else:\n",
    "            res = [exon_1[\"Chr\"],exon_1[\"Start\"]-3,exon_1[\"Start\"]-1,exon_1[\"Strand\"],exon_1[\"Transcript_id\"],\"3\"]\n",
    "        splice_list.append(res)\n",
    "        # middle exon\n",
    "        if len(df_i) > 2:\n",
    "            for i in range(1, len(df_i)-1):\n",
    "                exon_m = df_i.iloc[i, :]\n",
    "                if exon_1[\"Strand\"] == \"+\":\n",
    "                    splice_3=[exon_m[\"Chr\"],exon_m[\"Start\"]-3,exon_m[\"Start\"]-1,exon_m[\"Strand\"],exon_m[\"Transcript_id\"],\"3\"]\n",
    "                    splice_5=[exon_m[\"Chr\"],exon_m[\"End\"],exon_m[\"End\"]+2,exon_m[\"Strand\"],exon_m[\"Transcript_id\"],\"5\"]\n",
    "                else:\n",
    "                    splice_3=[exon_m[\"Chr\"],exon_m[\"End\"],exon_m[\"End\"]+2,exon_m[\"Strand\"],exon_m[\"Transcript_id\"],\"5\"]\n",
    "                    splice_5=[exon_m[\"Chr\"],exon_m[\"Start\"]-3,exon_m[\"Start\"]-1,exon_m[\"Strand\"],exon_m[\"Transcript_id\"],\"3\"]\n",
    "                splice_list.append(splice_3)\n",
    "                splice_list.append(splice_5)\n",
    "        # last exon\n",
    "        exon_f = df_i.iloc[-1, :]\n",
    "        if exon_1[\"Strand\"] == \"+\":\n",
    "            res = [exon_f[\"Chr\"],exon_f[\"Start\"]-3,exon_f[\"Start\"]-1,exon_f[\"Strand\"],exon_f[\"Transcript_id\"],\"3\"]\n",
    "        else:\n",
    "            res = [exon_f[\"Chr\"],exon_f[\"End\"],exon_f[\"End\"]+2,exon_f[\"Strand\"],exon_f[\"Transcript_id\"],\"5\"]\n",
    "        splice_list.append(res)\n",
    "    df_splice = pd.DataFrame(splice_list, columns=[\"Chr\", \"Start\", \"End\", \"Strand\", \"Transcript_id\", \"Event\"])\n",
    "    df_splice[\"Start\"] = df_splice[\"Start\"].astype(int)\n",
    "    df_splice[\"End\"] = df_splice[\"End\"].astype(int)\n",
    "    return df_splice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result[\"name\"]=df_result[\"Transcript_id\"]+\";\"+df_result[\"Chr\"]+\":\"+df_result[\"Start\"].astype(str)+\"-\"+df_result[\"End\"].astype(str)+\";\"+df_result[\"Event\"]\n",
    "df_result[\"dot\"] = \".\"\n",
    "df_final = df_result[[\"Chr\", \"Start\", \"End\", \"name\", \"dot\", \"Strand\"]]\n",
    "df_final.to_csv(\"/home/galaxy/data/splicing_sites/splice_sites.bed\", sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_fasta(\"/home/galaxy/data/splicing_sites/splice_sites.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta(bed):\n",
    "    result_file = \"/home/galaxy/data/splicing_sites/splice_sites.fa\"\n",
    "    reference_genome = \"/data/database/GRCh38/GENCODE/GRCh38.primary_assembly.genome.fa\"\n",
    "    command = \"bedtools getfasta -fi %s -bed %s -s -name\" % (reference_genome, bed)\n",
    "    sub_p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)\n",
    "    fa_sequence = str(sub_p.communicate()[0].decode(\"utf-8\"))\n",
    "    with open(result_file, 'w') as fw:\n",
    "        fw.write(fa_sequence)\n",
    "# return fa_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chr      start        end  \\\n",
      "0  chrX  100636605  100636607   \n",
      "1  chrX  100635746  100635748   \n",
      "2  chrX  100635555  100635557   \n",
      "3  chrX  100635252  100635254   \n",
      "4  chrX  100635175  100635177   \n",
      "\n",
      "                                                name dot strand Nucle  \n",
      "0  ENSG00000000003.14;ENST00000373020.8;chrX:1006...   .      -    GT  \n",
      "1  ENSG00000000003.14;ENST00000373020.8;chrX:1006...   .      -    AG  \n",
      "2  ENSG00000000003.14;ENST00000373020.8;chrX:1006...   .      -    GT  \n",
      "3  ENSG00000000003.14;ENST00000373020.8;chrX:1006...   .      -    AG  \n",
      "4  ENSG00000000003.14;ENST00000373020.8;chrX:1006...   .      -    GT  \n"
     ]
    }
   ],
   "source": [
    "# combine bed and sequence\n",
    "data_dir = \"/home/galaxy/data/splicing_sites\"\n",
    "bed = os.path.join(data_dir, \"splice_sites.bed\")\n",
    "fa = os.path.join(data_dir, \"splice_sites.fa\")\n",
    "result_file = os.path.join(data_dir, \"splice_sites.txt\")\n",
    "####################################################\n",
    "df_bed = pd.read_table(bed, sep=\"\\t\", header=None, names=[\"chr\", \"start\", \"end\", \"name\", \"dot\", \"strand\"])\n",
    "# print(df_bed.head())\n",
    "# df_fa = pd.read_table(fa, sep=\"\\t\", header=None)\n",
    "name_list, nucleo_list = [], []\n",
    "with open(fa, 'r') as f:\n",
    "    contents = f.readlines()\n",
    "    for i in range(0, len(contents), 2):\n",
    "        name_list.append(contents[i].split(\">\")[1].strip().split(\"(\")[0])\n",
    "        nucleo_list.append(contents[i+1].strip())\n",
    "df_fa = pd.DataFrame({\"name\": name_list, \"Nucle\": nucleo_list})\n",
    "# print(df_fa.head())\n",
    "df = df_bed.merge(df_fa, on=\"name\", how=\"left\")\n",
    "print(df.head())\n",
    "df.to_csv(result_file, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "## filter splice sites based on nucleotide\n",
    "## This is method needs to be check!!!!!!!!!!!!!!!!!!\n",
    "data_dir = \"/home/galaxy/data/splicing_sites/\"\n",
    "nucleo_file = os.path.join(data_dir, \"splice_sites.txt\")\n",
    "res_file = os.path.join(data_dir, \"filtered_splice_sites.txt\")\n",
    "###########################################################\n",
    "result_list = []\n",
    "with open(nucleo_file, 'r') as f:\n",
    "    title = f.readline()\n",
    "    contents = f.readlines()\n",
    "    for i in range(0, len(contents), 2):\n",
    "        if contents[i].strip().split(\"\\t\")[-1] == \"GT\" and contents[i+1].strip().split(\"\\t\")[-1] == \"AG\":\n",
    "            result_list.append(contents[i])\n",
    "            result_list.append(contents[i+1])\n",
    "with open(res_file, 'w') as fw:\n",
    "    result_list= [\"%s\\n\" % x.strip() for x in result_list]\n",
    "    fw.writelines(result_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "##############\n",
    "# 100bp upstream of 5SS, 100bp afterstream of 3SS\n",
    "#\n",
    "#                  up                down  \n",
    "# Strand: -, [end, end+100], [start-99, start+1]\n",
    "# Strand: +, [start-99, start+1], [end, end+100]\n",
    "# Consider intron\n",
    "# Strand: -, [end-100, end+100], [start-99, start+101]\n",
    "# Strand: +, [start-99, start+101], [end-100, end+100]\n",
    "##############\n",
    "sp_file = \"/home/galaxy/data/splicing_sites/filtered_splice_sites.txt\"\n",
    "result_file = \"/home/galaxy/data/splicing_sites/splice_sites_flank_withIntron.bed\"\n",
    "##############\n",
    "df_sp = pd.read_table(sp_file, sep=\"\\t\", names=[\"chr\", \"start\", \"end\", \"info\", \"dot\", \"strand\", \"nucleotide\"])\n",
    "df_sp[\"Type\"] = df_sp[\"info\"].str.split(\";\").str[-1]\n",
    "df_sp[\"info\"] = df_sp[\"info\"] + ';' + df_sp[\"nucleotide\"]\n",
    "# print(df_sp.head())\n",
    "#\n",
    "df_a = df_sp[(df_sp[\"Type\"]==\"5\") & (df_sp[\"strand\"]==\"+\")]\n",
    "df_a[\"new_start\"], df_a[\"new_end\"] = df_a[\"start\"]-99, df_a[\"start\"]+101\n",
    "df_b = df_sp[(df_sp[\"Type\"]==\"5\") & (df_sp[\"strand\"]==\"-\")]\n",
    "df_b[\"new_start\"], df_b[\"new_end\"] = df_b[\"end\"]-100, df_b[\"end\"]+100\n",
    "df_c = df_sp[(df_sp[\"Type\"]==\"3\") & (df_sp[\"strand\"]==\"+\")]\n",
    "df_c[\"new_start\"], df_c[\"new_end\"] = df_c[\"end\"]-100, df_c[\"end\"]+100\n",
    "df_d = df_sp[(df_sp[\"Type\"]==\"3\") & (df_sp[\"strand\"]==\"-\")]\n",
    "df_d[\"new_start\"], df_d[\"new_end\"] = df_d[\"start\"]-99, df_d[\"start\"]+101\n",
    "df_list = []\n",
    "for df in [df_a, df_b, df_c, df_d]:\n",
    "    df_list.append(df[[\"chr\", \"new_start\", \"new_end\", \"info\", \"dot\", \"strand\"]])\n",
    "df_bed = pd.concat(df_list)\n",
    "df_bed.to_csv(result_file, sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Density Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pre process has done!\n",
      "Brain_Nucleus_accumbens_basal_ganglia.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[325, 421, 490, 692, 301, 300, 286, 556, 537, 885]\n",
      "[250, 421, 365, 382, 345, 471, 442, 476, 376, 231]\n",
      "Heart_Left_Ventricle.v7.signif_variant_gene_pairs.bed-Heart.bed\n",
      "[386, 589, 486, 593, 459, 642, 461, 543, 568, 451]\n",
      "[416, 508, 529, 446, 441, 422, 686, 669, 580, 369]\n",
      "Brain_Substantia_nigra.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[55, 114, 154, 339, 97, 101, 108, 89, 171, 326]\n",
      "[58, 192, 92, 99, 119, 161, 116, 243, 92, 79]\n",
      "Brain_Hippocampus.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[175, 210, 251, 427, 145, 145, 208, 252, 353, 567]\n",
      "[151, 337, 198, 189, 158, 285, 268, 317, 195, 122]\n",
      "Heart_Atrial_Appendage.v7.signif_variant_gene_pairs.bed-Heart.bed\n",
      "[389, 534, 599, 754, 514, 610, 486, 656, 663, 588]\n",
      "[461, 556, 543, 447, 466, 592, 781, 693, 704, 406]\n",
      "Brain_Hypothalamus.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[160, 188, 272, 515, 184, 138, 219, 227, 347, 578]\n",
      "[123, 208, 212, 223, 145, 250, 263, 290, 253, 146]\n",
      "Adrenal_Gland.v7.signif_variant_gene_pairs.bed-Kidney.bed\n",
      "[355, 573, 483, 502, 432, 447, 500, 543, 684, 466]\n",
      "[413, 546, 491, 423, 451, 491, 463, 677, 484, 363]\n",
      "Brain_Frontal_Cortex_BA9.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[322, 379, 369, 691, 302, 307, 280, 390, 501, 758]\n",
      "[243, 491, 336, 294, 290, 399, 342, 405, 333, 226]\n",
      "Brain_Putamen_basal_ganglia.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[212, 306, 363, 587, 201, 245, 273, 334, 427, 637]\n",
      "[162, 359, 221, 255, 248, 404, 313, 397, 285, 211]\n",
      "Brain_Cerebellum.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[627, 930, 1143, 1386, 643, 689, 940, 1137, 1264, 1511]\n",
      "[739, 1030, 899, 957, 598, 1090, 978, 1003, 1077, 603]\n",
      "Muscle_Skeletal.v7.signif_variant_gene_pairs.bed-Muscle.bed\n",
      "[524, 922, 673, 740, 688, 780, 659, 1022, 1156, 903]\n",
      "[709, 708, 806, 777, 601, 758, 869, 900, 1020, 576]\n",
      "Brain_Cerebellar_Hemisphere.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[476, 649, 845, 1079, 515, 527, 686, 704, 834, 1225]\n",
      "[511, 714, 601, 613, 509, 719, 695, 811, 626, 377]\n",
      "Brain_Anterior_cingulate_cortex_BA24.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[274, 298, 340, 555, 224, 231, 214, 323, 468, 719]\n",
      "[204, 327, 266, 260, 228, 360, 337, 428, 296, 211]\n",
      "Stomach.v7.signif_variant_gene_pairs.bed-Stomach.bed\n",
      "[218, 260, 191, 257, 231, 319, 352, 374, 531, 327]\n",
      "[203, 348, 241, 266, 236, 271, 321, 396, 373, 222]\n",
      "Brain_Caudate_basal_ganglia.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[377, 513, 604, 713, 316, 365, 430, 563, 601, 772]\n",
      "[272, 477, 458, 509, 344, 468, 412, 559, 464, 306]\n",
      "Brain_Amygdala.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[103, 173, 173, 425, 102, 79, 139, 158, 251, 458]\n",
      "[116, 196, 128, 122, 102, 191, 165, 217, 139, 84]\n",
      "Pituitary.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[563, 421, 897, 1039, 462, 438, 528, 682, 829, 953]\n",
      "[418, 606, 489, 548, 415, 654, 764, 673, 618, 378]\n",
      "Liver.v7.signif_variant_gene_pairs.bed-Liver.bed\n",
      "[96, 112, 106, 159, 75, 97, 160, 128, 198, 85]\n",
      "[88, 117, 126, 89, 115, 95, 140, 119, 178, 94]\n",
      "Brain_Cortex.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[410, 519, 640, 794, 399, 472, 495, 545, 745, 896]\n",
      "[383, 629, 459, 534, 399, 619, 555, 753, 536, 347]\n",
      "Brain_Spinal_cord_cervical_c-1.v7.signif_variant_gene_pairs.bed-Brain.bed\n",
      "[98, 167, 191, 403, 119, 99, 160, 126, 211, 416]\n",
      "[104, 185, 128, 137, 101, 243, 171, 217, 123, 100]\n",
      "Lung.v7.signif_variant_gene_pairs.bed-Lung.bed\n",
      "[504, 676, 500, 519, 692, 692, 909, 886, 935, 1229]\n",
      "[653, 814, 807, 612, 554, 713, 745, 963, 906, 532]\n"
     ]
    }
   ],
   "source": [
    "####################################################################################\n",
    "eqtl_dir = \"/home/galaxy/data/QTL/m6a_related_SNPs/eQTL/m6a_qtl_bed/\"\n",
    "eqtl_list = glob.glob(\"%s/*.bed\" % eqtl_dir)\n",
    "flank_file = \"/home/galaxy/data/splicing_sites/splice_sites_flank_withIntron.bed\"\n",
    "result_dir = \"/home/galaxy/data/splicing_sites/density_analysis\"\n",
    "#################################################################################\n",
    "def distinguish_orientation():\n",
    "    df = pd.read_table(flank_file, sep=\"\\t\", header=None, names=[\"chr\", \"start\", \"end\", \"info\", \"dot\", \"strand\"])\n",
    "    df[\"Type\"] = df[\"info\"].str.split(\";\").str[-2]  # !\n",
    "    df_5, df_3 = df[df[\"Type\"]==\"5\"], df[df[\"Type\"]==\"3\"]\n",
    "    return df_5, df_3\n",
    "\n",
    "def split_into_bins(df):\n",
    "    n_arrays = []\n",
    "    for x,values in df.iterrows():\n",
    "        bin_list,chr_n,start,end,strand =[],values[\"chr\"],int(values[\"start\"]),int(values[\"end\"]),values[\"strand\"]\n",
    "        for i in range(start, end, 20):\n",
    "            bin_list.append([chr_n, i, i+20])\n",
    "        # reverse - strand order\n",
    "        if strand == \"+\":\n",
    "            n_arrays.append(bin_list)\n",
    "        else:\n",
    "            bin_list.reverse()\n",
    "            n_arrays.append(bin_list)\n",
    "    return n_arrays\n",
    "\n",
    "def transform_into_bed(n_arrays):\n",
    "    total_bins = []\n",
    "    for j in range(len(n_arrays[0])):\n",
    "        bin_str, bin_list = \"\", []\n",
    "        for i in range(len(n_arrays)):\n",
    "            i_list = n_arrays[i][j]\n",
    "            str_bed = \"\\t\".join([str(x) for x in i_list])\n",
    "            bin_list.append(str_bed)\n",
    "        # each bin\n",
    "        bin_str = \"\\n\".join(bin_list)\n",
    "        total_bins.append(bin_str)\n",
    "    return total_bins\n",
    "            \n",
    "\n",
    "def overlap_with_snp(eqtl, str_bin):\n",
    "    command = \"bedtools intersect -a stdin -b %s -wa | wc -l\" % eqtl\n",
    "    sub_p = subprocess.Popen(command, shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE)\n",
    "    str_bin = bytes(str_bin, 'utf-8')\n",
    "    snp_num = int(sub_p.communicate(str_bin)[0].decode(\"utf-8\").strip())\n",
    "    return snp_num\n",
    "    \n",
    "def main_method():\n",
    "    df_5, df_3 = distinguish_orientation()\n",
    "    array_5, array_3 = split_into_bins(df_5), split_into_bins(df_3)\n",
    "    str5_list, str3_list = transform_into_bed(array_5), transform_into_bed(array_3)\n",
    "    print(\"Pre process has done!\")\n",
    "    for eqtl in eqtl_list:\n",
    "        print(os.path.basename(eqtl))\n",
    "        str5_num = [overlap_with_snp(eqtl, str5) for str5 in str5_list]\n",
    "        print(str5_num)\n",
    "        str3_num = [overlap_with_snp(eqtl, str3) for str3 in str3_list]\n",
    "        print(str3_num)\n",
    "#         result_file_up = os.path.join(result_dir, \"up-%s\" % os.path.basename(eqtl))\n",
    "#         result_file_down = os.path.join(result_dir, \"down-%s\" % os.path.basename(eqtl))\n",
    "#         df_result_up.to_csv(result_file_up, sep=\"\\t\", header=False, index=False)\n",
    "#         df_result_down.to_csv(result_file_down, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "main_method() # eqtl_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0      151.0\n",
      "0.0.1    188.0\n",
      "0.0.2    270.0\n",
      "0.0.3    262.0\n",
      "0.0.4    257.0\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "result_dir = \"/home/galaxy/data/splicing_sites/density_analysis\"\n",
    "result_file_up = os.path.join(result_dir, \"down-Brain_Putamen_basal_ganglia.v7.signif_variant_gene_pairs.bed-Brain.bed\")\n",
    "result_file_down = os.path.join(result_dir, \"splice_sites_flank_down.bed\")\n",
    "df_up = pd.read_table(result_file_up, sep=\"\\t\")\n",
    "df_down = pd.read_table(result_file_down, sep=\"\\t\")\n",
    "# for x in df_up:\n",
    "#     print(sum(df_up[x]))\n",
    "print(df_up.sum(0))\n",
    "# print(\"###################\")\n",
    "# print(df_down.sum(0))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
