{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate match SNP sets\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig/bed/SNPsnap/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list = glob.glob(\"SNPsnap_*/matched_snps.txt\")\n",
    "for x in file_list:\n",
    "    result_file = x.replace(\"txt\",'bed')\n",
    "    with open(result_file, 'w') as fw:\n",
    "        with open(x, 'r') as f:\n",
    "            f.readline()\n",
    "            content = f.readlines()\n",
    "            for line in content:\n",
    "                snp_list = line.strip().split(\"\\t\")\n",
    "                query_snp, remain_snp = snp_list[0], snp_list[1:]\n",
    "                a = query_snp.split(\":\")[0] + \":\"\n",
    "                condition = 0\n",
    "                for y in remain_snp:\n",
    "                    if y.startswith(a):\n",
    "                        chr_y,pos_y = \"chr\" + y.split(\":\")[0], y.split(\":\")[1].strip()\n",
    "                        fw.write(\"%s\\t%s\\t%s\\n\" % (chr_y, pos_y, pos_y))\n",
    "                        condition = 1\n",
    "                        break\n",
    "                if condition == 0:\n",
    "                    y=remain_snp[0]\n",
    "                    chr_y,pos_y = \"chr\" + y.split(\":\")[0], y.split(\":\")[1].strip()\n",
    "                    fw.write(\"%s\\t%s\\t%s\\n\" % (chr_y, pos_y, pos_y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                       Disease trait                  ...                                             Parent URI\n",
      "0                                           Clubfoot                  ...                   http://www.ebi.ac.uk/efo/EFO_0000408\n",
      "1  Alcohol dependence or heroin dependence or met...                  ...                   http://www.ebi.ac.uk/efo/EFO_0000618\n",
      "2                         Methamphetamine dependence                  ...                   http://www.ebi.ac.uk/efo/EFO_0000618\n",
      "3                                     Hypothyroidism                  ...                   http://www.ebi.ac.uk/efo/EFO_0000408\n",
      "4                                   Age at menopause                  ...                   http://www.ebi.ac.uk/efo/EFO_0001444\n",
      "\n",
      "[5 rows x 5 columns]\n",
      "Index(['Disease trait', 'EFO term', 'EFO URI', 'Parent term', 'Parent URI'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# process cluster\n",
    "efo = \"/home/database/GWAS/gwas_catalog_trait-mappings_r2020-05-03.tsv\"\n",
    "df = pd.read_table(efo)\n",
    "print(df.head())\n",
    "print(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extend to 250bp\n",
    "def extend_bed(bed):\n",
    "    res = bed.replace(\".bed\",'_extended.bed')\n",
    "    df = pd.read_table(bed, header=None)\n",
    "    df['s'] = df.iloc[:,1] - 250\n",
    "    df['e'] = df.iloc[:,1] + 251\n",
    "    df['chr'] = df.iloc[:,0]\n",
    "    df = df[['chr','s','e']].drop_duplicates()\n",
    "    df.to_csv(res, sep=\"\\t\", header=False, index=False)\n",
    "    \n",
    "# control set\n",
    "file_list = glob.glob(\"SNPsnap_*/matched_snps.bed\")\n",
    "for x in file_list:\n",
    "    extend_bed(x)\n",
    "# m6a set\n",
    "file_list = glob.glob(\"../*.bed\")\n",
    "for x in file_list:\n",
    "    extend_bed(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRR055525.bed\t9.11921145659114e-05\t1.8277939008660402\t164\t54\n",
      "CRR042298.bed\t0.12282649912079449\t1.411764705882353\t66\t34\n",
      "kid_IP_4.bed\t1.2948818792760652e-07\t3.074121451670859\t86\t25\n",
      "CRR042294.bed\t8.076951587777668e-05\t2.6041088284286507\t70\t20\n",
      "brain_IP_1.bed\t0.011529160404608337\t1.536189069423929\t100\t56\n",
      "CRR073016.bed\t0.3953534938064962\t1.274511008704557\t47\t25\n",
      "mus_IP_5.bed\t1.5924089295216472e-08\t3.0180996551314068\t103\t30\n",
      "CRR042286.bed\t0.36558013163965697\t1.1874876183683982\t82\t47\n",
      "CRR042320.bed\t0.03691742782560466\t1.9001721170395869\t40\t15\n",
      "CRR042292.bed\t0.002182105548703262\t1.8830211585154353\t86\t31\n",
      "CRR055543.bed\t0.15358503164431253\t1.4968285131767727\t45\t19\n",
      "mus_IP_4.bed\t0.0005497132744664217\t3.2050284629981025\t37\t10\n",
      "CRR042282.bed\t0.000499886182956333\t2.0813308293140977\t87\t28\n",
      "CRR073020.bed\t0.04725751698560339\t2.33839897260274\t25\t8\n",
      "CRR042302.bed\t3.2563607571676624e-05\t2.8843373493975903\t63\t18\n",
      "CRR055539.bed\t0.047671976774004365\t1.9661616161616162\t34\t12\n",
      "heart_IP_1.bed\t2.0279786939987358e-07\t4.085849056603774\t61\t14\n",
      "pla_IP_2.bed\t4.914760026267585e-05\t1.9444799205335603\t118\t54\n",
      "CRR055563.bed\t0.05916667807582608\t2.3038884192730347\t23\t7\n",
      "CRR055537.bed\t0.11941663922765554\t1.442423773409689\t57\t28\n",
      "heart_IP_3.bed\t0.0018167321091473406\t2.003655303030303\t67\t30\n",
      "CRR055549.bed\t0.13520741035510064\t1.809968847352025\t30\t10\n",
      "CRR073018.bed\t0.18739414106591684\t1.375684007707129\t58\t30\n",
      "CRR055541.bed\t1.0\t1.0184686977386752\t23\t17\n",
      "CRR042304.bed\t0.001433214730242638\t1.5498518151943639\t190\t71\n",
      "CRR055547.bed\t0.015214500485892459\t1.8662172673061965\t61\t21\n",
      "lung_IP_5.bed\t1.1537692350191493e-06\t3.371251441753172\t67\t18\n",
      "sto_IP_5.bed\t9.794594606755287e-07\t3.6766604593420236\t61\t15\n",
      "CRR042308.bed\t0.00295150220918079\t2.263322506916877\t53\t17\n",
      "sto_IP_4.bed\t0.0035775381633279925\t1.8406565399123254\t69\t34\n",
      "brain_IP_2.bed\t0.001954546616950539\t2.0385403160644153\t65\t27\n",
      "CRR042316.bed\t2.3722528440836673e-05\t2.6186591186591186\t87\t22\n",
      "CRR055553.bed\t0.10981982412699744\t1.4579028175442965\t66\t29\n",
      "kid_IP_3.bed\t0.04922635437211339\t1.4718633418417435\t74\t44\n",
      "CRR055533.bed\t0.00670622745095811\t3.6506024096385543\t24\t5\n",
      "CRR055557.bed\t0.02258263878628632\t2.8328267477203646\t24\t6\n",
      "CRR042290.bed\t0.0006216104713812477\t2.069372994448123\t84\t29\n",
      "CRR055529.bed\t0.0006486098082825642\t1.709090909090909\t141\t55\n",
      "CRR055527.bed\t5.8443934370254884e-05\t1.8916202621733653\t139\t54\n",
      "liver_IP_2.bed\t4.254629570117871e-08\t2.7862326177939787\t107\t36\n",
      "CRR042278.bed\t0.02375739854337175\t1.3743173568718032\t161\t80\n",
      "pla_IP_6.bed\t2.0404501417699174e-05\t2.366604032450489\t85\t33\n",
      "CRR055531.bed\t0.01732157283525177\t1.6707763436717395\t84\t29\n",
      "liver_IP_1.bed\t3.491778479484846e-08\t3.454711650324082\t82\t22\n",
      "CRR055551.bed\t0.012616788232857877\t2.3099532398129594\t38\t12\n",
      "CRR055555.bed\t0.013847563276374599\t2.7790432801822322\t28\t7\n",
      "CRR042310.bed\t0.10586065154834254\t1.4945726762320648\t55\t25\n",
      "CRR042300.bed\t0.0005931174763616266\t2.190488047808765\t76\t24\n",
      "brain_IP_3.bed\t4.777447027930539e-07\t2.395883603542501\t117\t43\n",
      "CRR055561.bed\t0.0027763212235336187\t2.1821588963199896\t55\t19\n",
      "heart_IP_2.bed\t1.588110376134226e-06\t2.622635089152035\t89\t31\n",
      "liver_IP_3.bed\t9.047112372545337e-05\t2.344836377120608\t71\t28\n",
      "CRR042306.bed\t0.005132221625093319\t1.9223628691983123\t67\t25\n",
      "CRR042284.bed\t0.6816293319490643\t0.9408797901876207\t148\t88\n",
      "pla_IP_4.bed\t0.001410947704755415\t1.8151032847951236\t88\t44\n",
      "CRR042296.bed\t3.4989714734852305e-06\t2.6622130810366103\t97\t25\n",
      "CRR042312.bed\t7.758685471516235e-05\t2.6335365853658534\t70\t20\n",
      "lung_IP_4.bed\t6.953399797328641e-07\t2.5255619611783997\t103\t37\n",
      "CRR055545.bed\t0.004473125298755219\t1.8696194973674212\t79\t27\n",
      "kid_IP_2.bed\t9.053241926284359e-08\t2.282969832857725\t142\t55\n",
      "CRR042280.bed\t9.571573421894457e-07\t2.743885906404208\t100\t27\n",
      "CRR042288.bed\t0.00123783476803654\t3.622036874451273\t37\t6\n",
      "CRR055536.bed\t0.00029638198648309483\t2.455025033581634\t73\t19\n",
      "CRR055559.bed\t0.0830583955424064\t2.1347248576850095\t21\t7\n",
      "CRR042318.bed\t0.5290839902027232\t1.116404011461318\t133\t56\n",
      "CRR042314.bed\t0.03543338543865213\t1.6454757363848274\t60\t26\n"
     ]
    }
   ],
   "source": [
    "#### enrichment analysis\n",
    "# gwas_dir = \"/data5/galaxy/project/GWAS_db/04_class_snp/cluster_by_map/snp_classification/hg19/reformat_bed/\"\n",
    "# gwas_list = glob.glob(\"%s/*.bed\" % gwas_dir)\n",
    "gwas = \"/home/database/GWAS/gwas_hg19.bed2\"\n",
    "##\n",
    "query_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig/bed/\"\n",
    "query_list = [x for x in glob.glob(\"%s/*.bed\" % query_dir) if \"_extended\" not in x]\n",
    "match_dir=\"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig/bed/SNPsnap/\"\n",
    "\n",
    "for m6a in query_list:\n",
    "    sample = os.path.basename(m6a).split(\".bed\")[0]  #####\n",
    "    m6a_num = int(os.popen(\"wc -l %s\"%m6a).read().split()[0])\n",
    "    con = match_dir + os.sep + \"SNPsnap_%s/matched_snps.bed\" % sample # _extended\n",
    "    con_num = int(os.popen(\"wc -l %s\"%con).read().split()[0])\n",
    "    #for gwas in gwas_list:\n",
    "    overlap_m6a=os.popen(\"bedtools intersect -a %s -b %s -wa|sort -u|wc -l\" % (gwas,m6a)).read().split()[0]\n",
    "    overlap_con=os.popen(\"bedtools intersect -a %s -b %s -wa|sort -u|wc -l\" % (gwas,con)).read().split()[0]\n",
    "    a,b,c,d = int(overlap_m6a), int(overlap_con), m6a_num-int(overlap_m6a), con_num-int(overlap_con)\n",
    "    oddsratio,pvalue = stats.fisher_exact([[a,b], [c,d]])\n",
    "#     print(\"####          ASM   Control\")\n",
    "#     print(\"# GWAS         %d       %d  \" % (a,b))\n",
    "#     print(\"# Non-GWAS     %d       %d  \" % (c,d))\n",
    "#     print(\"#############################\\n\\n\")\n",
    "    print(\"\\t\".join([os.path.basename(m6a), str(pvalue), str(oddsratio), str(a), str(b)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs284860\n",
      "rs284860\n",
      "rs2478839\n",
      "rs17731\n",
      "rs943037\n",
      "rs284860\n",
      "rs2478839\n",
      "rs943037\n",
      "rs10786736\n",
      "rs3739998\n",
      "rs11190141\n",
      "rs4074536\n",
      "rs4917833\n",
      "rs284860\n",
      "rs10786736\n",
      "rs10786736\n",
      "rs1127687\n",
      "rs6892\n",
      "rs284860\n",
      "rs10786736\n",
      "rs1127687\n",
      "rs10786736\n",
      "rs284860\n",
      "rs10786736\n",
      "rs284860\n",
      "rs10786736\n",
      "rs2478839\n",
      "rs6586145\n",
      "rs7332\n",
      "rs7908745\n",
      "rs284860\n",
      "rs10786736\n",
      "rs3739998\n",
      "rs1127687\n",
      "rs284860\n",
      "rs7908745\n",
      "rs284860\n",
      "rs284860\n",
      "rs284860\n",
      "rs284860\n",
      "rs284860\n",
      "rs2296887\n",
      "rs284860\n",
      "rs2297991\n",
      "rs1061810\n",
      "rs17731\n",
      "rs284860\n",
      "rs2297991\n",
      "rs10786736\n",
      "rs284860\n",
      "rs2478839\n",
      "rs1127687\n",
      "rs1801253\n",
      "rs10786736\n",
      "rs943037\n",
      "rs2296887\n",
      "rs1129555\n",
      "rs2478839\n",
      "rs1127687\n",
      "rs284860\n",
      "rs2296887\n",
      "rs17731\n",
      "rs284860\n",
      "rs10786736\n",
      "rs284860\n",
      "rs943037\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# allele-specific m6A\n",
    "os.chdir(\"/home/galaxy/project/alleleSpecific_analysis/results/ASE_analysis/readCount/Fisher_test/sig/bed/\")\n",
    "result_dir = \"/home/galaxy/project/alleleSpecific_analysis/results/GWAS/\"\n",
    "#\n",
    "gwas = \"/home/database/GWAS/gwas_hg19.bed2\"\n",
    "disease = \"/home/database/GWAS/gwas_disease.txt\"\n",
    "df = pd.read_table(disease)\n",
    "rs_dict = dict(zip(df['SNPS'],df['DISEASE/TRAIT']))\n",
    "# gwas_dir = \"/data5/galaxy/project/GWAS_db/04_class_snp/cluster_by_map/snp_classification/hg19/reformat_bed/\"\n",
    "# gwas_list = glob.glob(\"%s/*.bed\" % gwas_dir)\n",
    "m6a_list = [x for x in glob.glob(\"*.bed\") if \"_extended\" not in x]\n",
    "for m6a in m6a_list:#\n",
    "    # print(\"############################\")\n",
    "    #for gwas in gwas_list:\n",
    "#     df = pd.read_table(m6a, sep=\"\\t\", header=None)\n",
    "#     df['m6a'] = df.iloc[:,0] + \":\" + df.iloc[:,2].astype(str)\n",
    "#     m6a_list = df['m6a'].tolist()\n",
    "    # overlap_m6a=os.popen(\"bedtools intersect -a %s -b %s -wa|sort -u|wc -l\" % (m6a,gwas)).read().split()[0]\n",
    "    tmp_file = os.path.join(result_dir, 'tmp.txt')\n",
    "    os.system(\"bedtools intersect -a %s -b %s -wa|sort -u > %s\" % (gwas,m6a,tmp_file))\n",
    "    rs_list = pd.read_table(tmp_file,header=None).iloc[:,3].tolist()\n",
    "    result_file = os.path.join(result_dir, os.path.basename(m6a))\n",
    "    with open(result_file, 'w') as fw:\n",
    "        lines = [\"%s\\t%s\\n\" % (rs, rs_dict[rs]) for rs in rs_list if rs in rs_dict]\n",
    "        fw.writelines(lines)\n",
    "os.system(\"rm %s\" % tmp_file)\n",
    "    # print(os.path.basename(m6a), os.path.basename(gwas), overlap_m6a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (11,12,23,27,30) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  del sys.path[0]\n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  from ipykernel import kernelapp as app\n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  app.launch_new_instance()\n",
      "/home/galaxy/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "#####  pre-process gwas raw\n",
    "os.chdir(\"/home/database/GWAS/\")\n",
    "df = pd.read_table(\"gwas.tsv\")\n",
    "# cols = ['CHR_ID', 'CHR_POS','CHR_POS','SNPS','P-VALUE','DISEASE/TRAIT','REPORTED GENE(S)']\n",
    "# df_sub = df[cols]\n",
    "# df_sub.columns = ['CHR_ID', 'CHR_S','CHR_E','SNPS','P-VALUE','DISEASE/TRAIT','REPORTED GENE(S)']\n",
    "# df_sub['term'] = df['SNPS']+\";\"+df['P-VALUE'].astype(str)+';'+df['DISEASE/TRAIT']+';'+df['REPORTED GENE(S)']\n",
    "# df_res = df_sub[['CHR_ID','CHR_S','CHR_E','term']]\n",
    "# df_res.to_csv(\"GWAS_hg38.bed\", sep=\"\\t\", header=False, index=False)\n",
    "df['CHR_ID'] = \"chr\" + df['CHR_ID'].astype(str)\n",
    "df_sub = df[['CHR_ID', 'CHR_POS','CHR_POS','SNPS','DISEASE/TRAIT']]\n",
    "df_sub.columns = ['CHR_ID','CHR_S','CHR_E','SNPS','DISEASE/TRAIT']\n",
    "df_sub['CHR_ID'] = df_sub['CHR_ID'].str.split(\";\").str[0]\n",
    "df_sub['CHR_S'] = df_sub['CHR_S'].str.split(\";\").str[0]\n",
    "df_sub['CHR_E'] = df_sub['CHR_E'].str.split(\";\").str[0]\n",
    "df_sub['SNPS'] = df_sub['SNPS'].str.split(\";\").str[0]\n",
    "df_sub['DISEASE/TRAIT'] = df_sub['DISEASE/TRAIT'].str.split(\";\").str[0]\n",
    "df_sub = df_sub.dropna()\n",
    "df_sub['CHR_ID'] = df_sub['CHR_ID'].str.split(\" x\").str[0]\n",
    "df_sub['CHR_S'] = df_sub['CHR_S'].str.split(\" x\").str[0]\n",
    "df_sub['CHR_E'] = df_sub['CHR_E'].str.split(\" x\").str[0]\n",
    "df_sub['SNPS'] = df_sub['SNPS'].str.split(\" x\").str[0]\n",
    "df_sub['DISEASE/TRAIT'] = df_sub['DISEASE/TRAIT'].str.split(\" x\").str[0]\n",
    "# df_sub.to_csv(\"gwas_hg38.bed\", sep=\"\\t\", header=False, index=False)\n",
    "df_sub = df_sub[~df_sub['SNPS'].str.contains(',')]\n",
    "df_sub['CHR_S'] = df_sub['CHR_S'].astype(int) - 1\n",
    "df_map = df_sub[['SNPS', 'DISEASE/TRAIT']]\n",
    "df_map.to_csv(\"gwas_disease.txt\", sep=\"\\t\", index=False)\n",
    "df_sub.to_csv(\"gwas_hg38.bed\", sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
