{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N E X T F L O W  ~  version 18.10.1\n",
      "Launching `main.nf` [ecstatic_nobel] - revision: fd5653f043\n",
      "\u001b[33m ▄▄▄▄▄▄▄▄▄▄▄  ▄         ▄  ▄▄▄▄▄▄▄▄▄▄▄  ▄         ▄  ▄▄▄▄▄▄▄▄▄▄▄  ▄▄▄▄▄▄▄▄▄▄▄ \n",
      "\u001b[0m\u001b[33m▐░░░░░░░░░░░▌▐░▌       ▐░▌▐░░░░░░░░░░░▌▐░▌       ▐░▌▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌\n",
      "\u001b[0m\u001b[33m▐░█▀▀▀▀▀▀▀▀▀ ▐░▌       ▐░▌▐░█▀▀▀▀▀▀▀▀▀ ▐░▌       ▐░▌▐░█▀▀▀▀▀▀▀▀▀ ▐░█▀▀▀▀▀▀▀▀▀ \n",
      "\u001b[0m\u001b[33m▐░▌          ▐░▌       ▐░▌▐░▌          ▐░▌       ▐░▌▐░▌          ▐░▌          \n",
      "\u001b[0m\u001b[33m▐░█▄▄▄▄▄▄▄▄▄ ▐░█▄▄▄▄▄▄▄█░▌▐░█▄▄▄▄▄▄▄▄▄ ▐░▌       ▐░▌▐░▌          ▐░▌          \n",
      "\u001b[0m\u001b[33m▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌▐░▌       ▐░▌▐░▌          ▐░▌          \n",
      "\u001b[0m\u001b[33m ▀▀▀▀▀▀▀▀▀█░▌ ▀▀▀▀█░█▀▀▀▀  ▀▀▀▀▀▀▀▀▀█░▌▐░▌       ▐░▌▐░▌          ▐░▌          \n",
      "\u001b[0m\u001b[33m          ▐░▌     ▐░▌               ▐░▌▐░▌       ▐░▌▐░▌          ▐░▌          \n",
      "\u001b[0m\u001b[33m ▄▄▄▄▄▄▄▄▄█░▌     ▐░▌      ▄▄▄▄▄▄▄▄▄█░▌▐░█▄▄▄▄▄▄▄█░▌▐░█▄▄▄▄▄▄▄▄▄ ▐░█▄▄▄▄▄▄▄▄▄ \n",
      "\u001b[0m\u001b[33m▐░░░░░░░░░░░▌     ▐░▌     ▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌▐░░░░░░░░░░░▌\n",
      "\u001b[0m\u001b[33m ▀▀▀▀▀▀▀▀▀▀▀       ▀       ▀▀▀▀▀▀▀▀▀▀▀  ▀▀▀▀▀▀▀▀▀▀▀  ▀▀▀▀▀▀▀▀▀▀▀  ▀▀▀▀▀▀▀▀▀▀▀ \n",
      "\u001b[0mnull\n",
      "\u001b[35m============You are running m6APipe with the following parameters===============\u001b[0m\n",
      "\u001b[35mChecking parameters ...\u001b[0m\n",
      "\u001b[33m=====================================Reads types================================\u001b[0m\n",
      "\u001b[33mSingleEnd :                     \u001b[0m\u001b[32mfalse\u001b[0m\n",
      "\u001b[33mStranded :                      \u001b[0m\u001b[32mno\u001b[0m\n",
      "\u001b[33m====================================Mode selected==============================\u001b[0m\n",
      "\u001b[33maligners :                      \u001b[0m\u001b[32mstar\u001b[0m\n",
      "\u001b[33mpeakCalling_mode :              \u001b[0m\u001b[32mindependence\u001b[0m\n",
      "\u001b[33mexpression_analysis_mode :      \u001b[0m\u001b[32mnone\u001b[0m\n",
      "\u001b[33mmethylation_analysis_mode :     \u001b[0m\u001b[32mWilcox-test\u001b[0m\n",
      "\u001b[33m==================================Input files selected==========================\u001b[0m\n",
      "\u001b[33mReads Path:                     \u001b[0m\u001b[32m/home/galaxy/project/m6AQTL/2019_10_10/transcriptome_snp/SNPiR/shell/HeLa_cell/data/PE\u001b[0m\n",
      "\u001b[33mfasta file :                    \u001b[0m\u001b[32m/home/galaxy/project/m6AQTL/2019_10_10/transcriptome_snp/SNPiR/shell/SNPiR-master/data/formal_chr/hg38.fa\u001b[0m\n",
      "\u001b[33mGtf file :                      \u001b[0m\u001b[32m/data/database/GRCh38/GENCODE/gencode.v27.annotation.gtf\u001b[0m\n",
      "\u001b[33mDesign file :                   \u001b[0m\u001b[32mdesignfile_HeLa_PE.txt\u001b[0m\n",
      "\u001b[33mCompare file :                  \u001b[0m\u001b[32mfalse\u001b[0m\n",
      "\u001b[33mchromsizes file :               \u001b[0m\u001b[32m/data/database/hg38/hg38.chrom.sizes\u001b[0m\n",
      "\u001b[33m==================================Skip model selected==========================\u001b[0m\n",
      "\u001b[33mSkip samtools sort :            \u001b[0m\u001b[32mfalse\u001b[0m\n",
      "\u001b[33mSkip expression analysis :      \u001b[0m\u001b[32mtrue\u001b[0m\n",
      "\u001b[33mSkip peakCalling :              \u001b[0m\u001b[32mfalse\u001b[0m\n",
      "\u001b[33mSkip diffpeakCalling :          \u001b[0m\u001b[32mtrue\u001b[0m\n",
      "\u001b[33mSkip annotation :               \u001b[0m\u001b[32mtrue\u001b[0m\n",
      "\u001b[33mSkip qc :                       \u001b[0m\u001b[32mfalse\u001b[0m\n",
      "\u001b[33m==================================Output files directory========================\u001b[0m\n",
      "\u001b[33mOutput directory :              \u001b[0m\u001b[32m/home/galaxy/project/m6AQTL/2019_10_10/transcriptome_snp/SNPiR/shell/HeLa_cell/results/PE/results\u001b[0m\n",
      "[warm up] executor > local\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[26/cbdf4b] Cached process > makeBED12 (gtf2bed12)\n",
      "[4e/1fbea0] Submitted process > Fastp (SRR5925299)\n",
      "[72/efd6bd] Submitted process > Fastp (SRR6714646)\n",
      "[69/b99a38] Submitted process > Fastp (SRR6714640)\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[b0/1c5101] Submitted process > Fastp (SRR5925273)\n",
      "\u001b[35mInitial reads mapping of SRR6714646 performed by STAR in paired-end mode\u001b[0m\n",
      "[26/d123e7] Submitted process > Fastp (SRR6714642)\n",
      "\u001b[35mInitial reads mapping of SRR5925299 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[59/7471bf] Submitted process > Fastp (SRR5925274)\n",
      "\u001b[35mInitial reads mapping of SRR6714640 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[cc/3d8e0d] Submitted process > Fastp (SRR5925301)\n",
      "\u001b[35mInitial reads mapping of SRR5925273 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[cc/f48790] Submitted process > Fastp (SRR6714644)\n",
      "\u001b[35mInitial reads mapping of SRR5925274 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[de/d032a5] Submitted process > Fastp (SRR6714643)\n",
      "\u001b[35mInitial reads mapping of SRR5925301 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[d2/f2defb] Submitted process > Fastp (SRR5925300)\n",
      "\u001b[35mInitial reads mapping of SRR6714642 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "\u001b[35mInitial reads mapping of SRR6714644 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mfastp is on going for pair-end data\u001b[0m\n",
      "[90/1f63a0] Submitted process > Fastp (SRR5925287)\n",
      "[82/8a5224] Submitted process > Fastp (SRR5925302)\n",
      "\u001b[35mInitial reads mapping of SRR5925287 performed by STAR in paired-end mode\u001b[0m\n",
      "[c6/eb6142] Submitted process > Fastp (SRR5925271)\n",
      "\u001b[35mInitial reads mapping of SRR6714643 performed by STAR in paired-end mode\u001b[0m\n",
      "[52/548e65] Submitted process > Fastp (SRR5925272)\n",
      "\u001b[35mInitial reads mapping of SRR5925300 performed by STAR in paired-end mode\u001b[0m\n",
      "[1d/f82a5e] Submitted process > Fastp (SRR5925289)\n",
      "\u001b[35mInitial reads mapping of SRR5925272 performed by STAR in paired-end mode\u001b[0m\n",
      "[1d/bf069b] Submitted process > Fastp (SRR6714641)\n",
      "[ab/213fa3] Submitted process > Fastp (SRR5925290)\n",
      "[7b/8575f1] Submitted process > Fastp (SRR5925288)\n",
      "[3a/8ce7db] Submitted process > Fastp (SRR6714647)\n",
      "[6d/671644] Submitted process > Fastp (SRR6714645)\n",
      "[2c/17ecdc] Submitted process > StarAlign (SRR6714646)\n",
      "[3e/43b6af] Submitted process > StarAlign (SRR5925299)\n",
      "[a0/623909] Submitted process > StarAlign (SRR6714640)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "\u001b[35mInitial reads mapping of SRR5925271 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mInitial reads mapping of SRR5925289 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[5f/b342ae] Submitted process > StarAlign (SRR5925273)\n",
      "\u001b[35mInitial reads mapping of SRR5925302 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[d5/aebe4d] Submitted process > StarAlign (SRR5925274)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[70/4937c4] Submitted process > StarAlign (SRR5925301)\n",
      "\u001b[35mInitial reads mapping of SRR5925290 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mInitial reads mapping of SRR6714641 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[61/079746] Submitted process > StarAlign (SRR6714642)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "\u001b[35mInitial reads mapping of SRR5925288 performed by STAR in paired-end mode\u001b[0m\n",
      "[80/a0124b] Submitted process > StarAlign (SRR6714644)\n",
      "\u001b[35mInitial reads mapping of SRR6714645 performed by STAR in paired-end mode\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[39/629278] Submitted process > StarAlign (SRR5925287)\n",
      "[01/1969ae] Submitted process > StarAlign (SRR6714643)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "\u001b[35mInitial reads mapping of SRR6714647 performed by STAR in paired-end mode\u001b[0m\n",
      "[0b/185d14] Submitted process > StarAlign (SRR5925300)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[d9/6b0029] Submitted process > StarAlign (SRR5925272)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[09/d78fb2] Submitted process > StarAlign (SRR5925271)\n",
      "[98/91a6c3] Submitted process > StarAlign (SRR5925289)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[8d/446885] Submitted process > StarAlign (SRR5925302)\n",
      "[52/0016fb] Submitted process > StarAlign (SRR5925290)\n",
      "[94/863f4f] Submitted process > StarAlign (SRR6714641)\n",
      "[dd/92374d] Submitted process > StarAlign (SRR5925288)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[a3/fdabb4] Submitted process > StarAlign (SRR6714645)\n",
      "[66/a09418] Submitted process > StarAlign (SRR6714647)\n",
      "[18/9e44f7] Submitted process > Sort (SRR6714646_star)\n",
      "[22/e07441] Submitted process > Sort (SRR5925299_star)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[5b/73b981] Submitted process > Sort (SRR6714640_star)\n",
      "[5d/bb2c9e] Submitted process > Sort (SRR5925273_star)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[fd/489298] Submitted process > Sort (SRR5925274_star)\n",
      "[0c/64ba92] Submitted process > Sort (SRR5925301_star)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[10/8a0a26] Submitted process > Sort (SRR6714644_star)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[c8/33fb8f] Submitted process > Sort (SRR6714642_star)\n",
      "[ea/b04b8c] Submitted process > Sort (SRR5925287_star)\n",
      "\u001b[35mSamtools sorting the bam files now\u001b[0m\n",
      "[34/d6fa82] Submitted process > Sort (SRR6714643_star)\n",
      "[a3/026a00] Submitted process > Sort (SRR5925300_star)\n",
      "[83/e84864] Submitted process > Sort (SRR5925272_star)\n",
      "[c8/26e427] Submitted process > Sort (SRR5925289_star)\n",
      "[92/e7222c] Submitted process > Sort (SRR5925271_star)\n",
      "[27/9f9bd5] Submitted process > Sort (SRR5925290_star)\n",
      "[56/00c153] Submitted process > Sort (SRR5925302_star)\n",
      "[10/a96d6a] Submitted process > Sort (SRR6714641_star)\n",
      "[ba/7779fd] Submitted process > Sort (SRR6714645_star)\n",
      "[b9/4270ab] Submitted process > Sort (SRR5925288_star)\n",
      "[11/1a61e9] Submitted process > Sort (SRR6714647_star)\n",
      "\u001b[35mRename the files for downstream analysis\u001b[0m\n",
      "[9e/6b758a] Submitted process > RenameByDesignfile\n",
      "\u001b[35mPeak Calling performed by Macs2 in independent mode\u001b[0m\n",
      "\u001b[35mPeak Calling performed by Meyer in independent mode\u001b[0m\n",
      "\u001b[35mPeak Calling performed by MeTPeak in independent mode\u001b[0m\n",
      "[59/f1f81a] Submitted process > RSeQC\n",
      "[1e/ed5ed3] Submitted process > Metpeak\n",
      "[a0/6721df] Submitted process > Meyer\n",
      "[c9/cd9921] Submitted process > multiqc\n",
      "ERROR ~ Error executing process > 'multiqc'\n",
      "\n",
      "Caused by:\n",
      "  Process `multiqc` terminated with an error exit status (127)\n",
      "\n",
      "Command executed:\n",
      "\n",
      "  multiqc -o multiqc .\n",
      "\n",
      "Command exit status:\n",
      "  127\n",
      "\n",
      "Command output:\n",
      "  (empty)\n",
      "\n",
      "Command error:\n",
      "  .command.sh: line 2: multiqc: command not found\n",
      "\n",
      "Work dir:\n",
      "  /home/software/m6APipe/m6APipe/work/c9/cd99212344ff0b388d7a5a69bfe2be\n",
      "\n",
      "Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`\n",
      "\n",
      " -- Check '.nextflow.log' file for details\n",
      "Execution cancelled -- Finishing pending tasks before exit\n",
      "\u001b[33mOops... Pipeline execution stopped with the following message: \u001b[0m\u001b[31m/home/software/m6APipe/m6APipe/work/c9/cd99212344ff0b388d7a5a69bfe2be/.command.sh: line 2: multiqc: command not found\u001b[0m\n",
      "^C\n"
     ]
    }
   ],
   "source": [
    "sra_dir = \"/home/galaxy/project/m6AQTL/2019_10_10/transcriptome_snp/SNPiR/shell/HeLa_cell/data/PE\"\n",
    "os.chdir(sra_dir)\n",
    "\n",
    "sra_list = glob.glob(\"*\")\n",
    "undo_list = [x for x in sra_list if not os.path.exists(os.path.join(sra_dir, x.split(\".\")[0] + \"_1.fastq\"))]\n",
    "for x in undo_list:\n",
    "    if \"fastq\" not in x:\n",
    "        prefix = x.split(\".\")[0]\n",
    "        os.system(\"fasterq-dump --split-3 %s -e 20 -p -o %s.fastq\" % (x, prefix))\n",
    "#####\n",
    "os.chdir(\"/home/software/m6APipe/m6APipe\")\n",
    "!nextflow main.nf -c nextflow.config --readPaths \"/home/galaxy/project/m6AQTL/2019_10_10/transcriptome_snp/SNPiR/shell/HeLa_cell/data/PE\" --designfile designfile_HeLa_PE.txt -profile standard -resume"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
